// @vercel/rust/dist/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  build: () => build,
  diagnostics: () => diagnostics,
  prepareCache: () => prepareCache,
  shouldServe: () => shouldServe,
  startDevServer: () => startDevServer2,
  version: () => version
});
module.exports = __toCommonJS(src_exports);
var import_node_path4 = __toESM(require("path"));
var import_build_utils7 = require("@vercel/build-utils");
var import_execa4 = __toESM(require("execa"));

// src/lib/rust-toolchain.ts
var import_build_utils = require("@vercel/build-utils");
var import_execa = __toESM(require("execa"));
async function downloadRustToolchain() {
  try {
    await (0, import_execa.default)(
      `curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- -y`,
      [],
      { shell: true, stdio: "inherit" }
    );
  } catch (err) {
    let message = "Unknown Error";
    if (err instanceof Error) {
      message = err.message;
    }
    throw new Error(`Installing Rust Toolchain via rustup failed: ${message}`);
  }
}
var installRustToolchain = async () => {
  try {
    await (0, import_execa.default)(`rustup -V`, [], { shell: true, stdio: "ignore" });
    (0, import_build_utils.debug)("Rust Toolchain is already installed, skipping download");
  } catch (_err) {
    await downloadRustToolchain();
  }
};

// src/lib/cargo.ts
var import_promises = require("fs/promises");
var import_node_fs = require("fs");
var import_node_path = __toESM(require("path"));
var import_smol_toml = require("smol-toml");
var import_execa2 = __toESM(require("execa"));
async function getCargoMetadata(options, filterPlatform) {
  const args = ["metadata", "--format-version", "1"];
  if (filterPlatform)
    args.push("--filter-platform", filterPlatform);
  const { stdout: cargoMetaData } = await (0, import_execa2.default)("cargo", args, options);
  return JSON.parse(cargoMetaData);
}
async function findCargoWorkspace(config) {
  const { stdout: projectDescriptionStr } = await (0, import_execa2.default)(
    "cargo",
    ["locate-project"],
    config
  );
  const projectDescription = JSON.parse(projectDescriptionStr);
  return {
    toml: (0, import_smol_toml.parse)(await (0, import_promises.readFile)(projectDescription.root, "utf8")),
    root: projectDescription.root
  };
}
async function findCargoBuildConfiguration(workspace) {
  const configPath = import_node_path.default.join(
    import_node_path.default.dirname(workspace.root),
    ".cargo/config.toml"
  );
  if (!(0, import_node_fs.existsSync)(configPath)) {
    return null;
  }
  const config = (0, import_smol_toml.parse)(await (0, import_promises.readFile)(configPath, "utf8"));
  return config;
}
function findBinaryName(workspace, entryPath) {
  const { bin } = workspace.toml;
  if (bin) {
    const relativePath = import_node_path.default.relative(import_node_path.default.dirname(workspace.root), entryPath);
    const entry = bin.find((binEntry) => binEntry.path === relativePath);
    if (entry?.name) {
      return entry.name;
    }
  }
  return import_node_path.default.basename(entryPath, ".rs").replace("[", "_").replace("]", "_");
}

// src/lib/utils.ts
var import_node_fs2 = __toESM(require("fs"));
var import_node_path2 = __toESM(require("path"));
var import_build_utils2 = require("@vercel/build-utils");
function getExecutableName(binName) {
  return process.platform === "win32" ? `${binName}.exe` : binName;
}
function assertEnv(name) {
  if (!process.env[name]) {
    throw new Error(`Missing ENV variable process.env.${name}`);
  }
  return process.env[name];
}
async function runUserScripts(dir) {
  const buildScriptPath = import_node_path2.default.join(dir, "build.sh");
  const buildScriptExists = import_node_fs2.default.existsSync(buildScriptPath);
  if (buildScriptExists) {
    (0, import_build_utils2.debug)("Running `build.sh`");
    await (0, import_build_utils2.runShellScript)(buildScriptPath);
  }
}
async function gatherExtraFiles(globMatcher, workPath) {
  if (!globMatcher)
    return {};
  (0, import_build_utils2.debug)(
    `Gathering extra files for glob \`${JSON.stringify(
      globMatcher
    )}\` in ${workPath}`
  );
  if (Array.isArray(globMatcher)) {
    const allMatches = await Promise.all(
      globMatcher.map((pattern) => (0, import_build_utils2.glob)(pattern, workPath))
    );
    return allMatches.reduce((acc, matches) => ({ ...acc, ...matches }), {});
  }
  return (0, import_build_utils2.glob)(globMatcher, workPath);
}

// src/lib/start-dev-server.ts
var import_child_process = require("child_process");
var import_events = require("events");
var import_get_port = __toESM(require("get-port"));
var import_build_utils5 = require("@vercel/build-utils");

// src/lib/dev-build.ts
var import_node_path3 = __toESM(require("path"));
var import_execa3 = __toESM(require("execa"));
var import_build_utils3 = require("@vercel/build-utils");
async function buildExecutableForDev(workPath, entrypoint) {
  (0, import_build_utils3.debug)(`Building executable for development: ${entrypoint}`);
  const HOME = process.platform === "win32" ? assertEnv("USERPROFILE") : assertEnv("HOME");
  const PATH = assertEnv("PATH");
  const rustEnv = {
    PATH: `${import_node_path3.default.join(HOME, ".cargo/bin")}${import_node_path3.default.delimiter}${PATH}`,
    RUSTFLAGS: [process.env.RUSTFLAGS].filter(Boolean).join(" ")
  };
  const entryPath = import_node_path3.default.join(workPath, entrypoint);
  const cargoWorkspace = await findCargoWorkspace({
    env: rustEnv,
    cwd: import_node_path3.default.dirname(entryPath)
  });
  const binaryName = findBinaryName(cargoWorkspace, entryPath);
  (0, import_build_utils3.debug)(`Building binary "${binaryName}" in debug mode for dev server`);
  try {
    await (0, import_execa3.default)(
      "cargo",
      [
        "build",
        "--bin",
        binaryName,
        "--message-format=json-diagnostic-rendered-ansi"
      ],
      {
        cwd: workPath,
        env: rustEnv,
        stdio: "pipe"
      }
    );
  } catch (err) {
    (0, import_build_utils3.debug)(`Cargo build failed for ${binaryName}`);
    throw new Error(`Failed to build Rust binary for development: ${err}`);
  }
  const { target_directory: targetDirectory } = await getCargoMetadata({
    cwd: workPath,
    env: rustEnv
  });
  const executablePath = import_node_path3.default.join(
    targetDirectory,
    "debug",
    getExecutableName(binaryName)
  );
  (0, import_build_utils3.debug)(`Built executable at: ${executablePath}`);
  return executablePath;
}

// src/lib/dev-server.ts
var import_build_utils4 = require("@vercel/build-utils");
function createDevServerEnv(baseEnv, meta = {}, port) {
  const devEnv = {
    // Base environment
    ...Object.fromEntries(
      Object.entries(baseEnv).filter(([, value]) => value !== void 0)
    ),
    // Development-specific variables
    VERCEL_DEV: "1",
    RUST_LOG: process.env.RUST_LOG || "info",
    // Runtime environment from meta
    ...meta.env || {}
  };
  if (typeof port === "number" && Number.isInteger(port)) {
    devEnv.VERCEL_DEV_PORT = String(port);
  }
  Object.keys(devEnv).forEach((key) => {
    if (devEnv[key] === void 0) {
      delete devEnv[key];
    }
  });
  (0, import_build_utils4.debug)(`Dev server environment: ${Object.keys(devEnv).join(", ")}`);
  return devEnv;
}

// src/lib/start-dev-server.ts
var SHUTDOWN_TIMEOUT = 35e3;
var ADDR_IN_USE_RE = /address (already )?in use|AddrInUse|EADDRINUSE/i;
var MAX_STDERR_CAPTURE = 8192;
var RUNNING_DEV_SERVERS = /* @__PURE__ */ new Set();
var cleanupHandlersInstalled = false;
function installGlobalCleanupHandlers() {
  if (cleanupHandlersInstalled)
    return;
  cleanupHandlersInstalled = true;
  const onSignal = () => {
    for (const child of RUNNING_DEV_SERVERS) {
      try {
        child.kill("SIGTERM");
      } catch (err) {
        (0, import_build_utils5.debug)(`Error sending SIGTERM to Rust dev server on signal: ${err}`);
      }
    }
  };
  process.on("SIGINT", onSignal);
  process.on("SIGTERM", onSignal);
  process.on("SIGHUP", onSignal);
  process.on("exit", () => {
    for (const child of RUNNING_DEV_SERVERS) {
      if (child.pid) {
        try {
          process.kill(child.pid, "SIGKILL");
        } catch {
        }
      }
    }
    RUNNING_DEV_SERVERS.clear();
  });
}
function trackDevServer(child) {
  installGlobalCleanupHandlers();
  RUNNING_DEV_SERVERS.add(child);
  const untrack = () => {
    RUNNING_DEV_SERVERS.delete(child);
  };
  child.once("exit", untrack);
  child.once("close", untrack);
}
var RustDevServerError = class extends Error {
  constructor(message) {
    super(message);
    this.name = "RustDevServerError";
  }
};
function terminate(child) {
  return new Promise((resolve) => {
    if (!child.pid || child.exitCode !== null || child.signalCode !== null) {
      resolve();
      return;
    }
    let timer;
    let settled = false;
    const done = () => {
      if (settled)
        return;
      settled = true;
      if (timer)
        clearTimeout(timer);
      resolve();
    };
    child.once("exit", done);
    child.once("close", done);
    try {
      child.kill("SIGTERM");
    } catch (err) {
      (0, import_build_utils5.debug)(`Error sending SIGTERM to Rust dev server: ${err}`);
      done();
      return;
    }
    timer = setTimeout(() => {
      (0, import_build_utils5.debug)(
        `Rust dev server did not exit within ${SHUTDOWN_TIMEOUT}ms, sending SIGKILL`
      );
      try {
        child.kill("SIGKILL");
      } catch (err) {
        (0, import_build_utils5.debug)(`Error sending SIGKILL to Rust dev server: ${err}`);
      }
    }, SHUTDOWN_TIMEOUT);
    timer.unref?.();
  });
}
var startDevServer = async (opts) => {
  const { entrypoint, workPath, meta = {}, onStdout, onStderr } = opts;
  try {
    await installRustToolchain();
    const executablePath = await buildExecutableForDev(workPath, entrypoint);
    const requestedPort = typeof meta.port === "number" ? meta.port : meta.env?.VERCEL_DEV_PORT ? Number(meta.env.VERCEL_DEV_PORT) : void 0;
    const port = typeof requestedPort === "number" && Number.isInteger(requestedPort) ? requestedPort : await (0, import_get_port.default)();
    (0, import_build_utils5.debug)(`Starting Rust dev server: ${executablePath} (port=${port})`);
    const devEnv = createDevServerEnv(process.env, meta, port);
    const child = (0, import_child_process.spawn)(executablePath, [], {
      cwd: workPath,
      env: devEnv,
      stdio: ["pipe", "pipe", "pipe"]
    });
    if (!child.pid) {
      throw new Error("Failed to start Rust dev server process");
    }
    trackDevServer(child);
    (0, import_build_utils5.debug)(`Rust dev server process started with PID: ${child.pid}`);
    let buffer = "";
    let portEmitted = false;
    let stderrTail = "";
    child.stdout?.on("data", (data) => {
      const chunk = Buffer.isBuffer(data) ? data : Buffer.from(data);
      buffer += chunk.toString();
      if (!portEmitted) {
        const match = buffer.match(/Dev server listening:\s*(\d+)/);
        if (match) {
          portEmitted = true;
          const reportedPort = parseInt(match[1], 10);
          (0, import_build_utils5.debug)(`Rust dev server reported ready on port ${reportedPort}`);
          child.emit("message", { port: reportedPort }, null);
          buffer = "";
        }
      }
      if (onStdout) {
        onStdout(chunk);
      } else {
        process.stdout.write(chunk.toString());
      }
    });
    child.stderr?.on("data", (data) => {
      const chunk = Buffer.isBuffer(data) ? data : Buffer.from(data);
      stderrTail = (stderrTail + chunk.toString()).slice(-MAX_STDERR_CAPTURE);
      if (onStderr) {
        onStderr(chunk);
      } else {
        process.stderr.write(chunk.toString());
      }
    });
    child.on("error", (err) => {
      (0, import_build_utils5.debug)(`Rust dev server error: ${err}`);
    });
    child.on("exit", (code, signal2) => {
      (0, import_build_utils5.debug)(`Rust dev server exited with code ${code}, signal ${signal2}`);
    });
    const onMessage = (0, import_events.once)(child, "message");
    const onExit = (0, import_events.once)(child, "close");
    const result = await Promise.race([
      onMessage.then((args) => {
        const [messageData] = args;
        return { state: "message", value: messageData };
      }),
      onExit.then((args) => {
        const [code, signal2] = args;
        return { state: "exit", value: [code, signal2] };
      })
    ]);
    if (result.state === "message") {
      const readyPort = typeof result.value?.port === "number" ? result.value.port : port;
      (0, import_build_utils5.debug)(`Rust dev server ready on port ${readyPort} (pid ${child.pid})`);
      if (!child.pid) {
        throw new Error("Child process has no PID");
      }
      return {
        port: readyPort,
        pid: child.pid,
        // Wait for exit so the port is released before `vercel dev` continues.
        shutdown: () => terminate(child)
      };
    }
    const [exitCode, signal] = result.value;
    const reason = signal ? `"${signal}" signal` : `exit code ${exitCode}`;
    const stderr = stderrTail.trim();
    if (ADDR_IN_USE_RE.test(stderr)) {
      throw new RustDevServerError(
        `Rust dev server failed to bind port ${port} ("address already in use"). A previous dev server instance may not have shut down yet. Please retry, or ensure no other process is using that port.`
      );
    }
    (0, import_build_utils5.debug)(
      `Rust dev server exited before becoming ready (${reason}). Falling back to build-and-invoke mode.` + (stderr ? ` stderr:
${stderr}` : "")
    );
    return null;
  } catch (error) {
    (0, import_build_utils5.debug)(`Failed to start Rust dev server: ${error}`);
    if (error instanceof RustDevServerError) {
      throw error;
    }
    return null;
  }
};

// src/diagnostics.ts
var import_build_utils6 = require("@vercel/build-utils");
function parseSource(source) {
  if (!source)
    return { include: false };
  if (source.startsWith("path+file:"))
    return { include: false };
  if (source.startsWith("registry+https://github.com/rust-lang/crates.io-index")) {
    return {
      include: true,
      source: "registry",
      sourceUrl: "https://crates.io"
    };
  }
  if (source.startsWith("registry+")) {
    const url = source.slice("registry+".length).split(/[?#]/)[0];
    return { include: true, source: "registry", sourceUrl: url };
  }
  if (source.startsWith("git+")) {
    const url = source.slice("git+".length).split(/[?#]/)[0];
    return { include: true, source: "git", sourceUrl: url };
  }
  return { include: true };
}
async function generateProjectManifest({
  workPath,
  cargoMetadata,
  framework,
  serviceType,
  runtimeVersion
}) {
  try {
    const { packages, resolve } = cargoMetadata;
    const pkgById = new Map(packages.map((p) => [p.id, p]));
    const rootId = resolve.root;
    const rootNode = resolve.nodes.find((n) => n.id === rootId);
    if (!rootNode)
      return;
    const rootPkg = pkgById.get(rootId);
    const directMap = /* @__PURE__ */ new Map();
    for (const dep of rootNode.deps) {
      const scopes = dep.dep_kinds.map(
        (dk) => dk.kind === "dev" ? "dev" : dk.kind === "build" ? "build" : "prod"
      );
      const depPkg = pkgById.get(dep.pkg);
      const req = rootPkg?.dependencies.find((d) => d.name === depPkg?.name)?.req;
      directMap.set(dep.pkg, {
        scopes: [...new Set(scopes)].sort(),
        requested: req
      });
    }
    const directEntries = [];
    const transitiveEntries = [];
    for (const node of resolve.nodes) {
      if (node.id === rootId)
        continue;
      const pkg = pkgById.get(node.id);
      if (!pkg)
        continue;
      const sourceInfo = parseSource(pkg.source);
      if (!sourceInfo.include)
        continue;
      const directInfo = directMap.get(node.id);
      const entry = {
        name: pkg.name,
        type: directInfo ? "direct" : "transitive",
        // Transitive scope would require full graph traversal to trace which
        // root-level scope pulled this in. 'prod' is a safe default — accurate
        // scope propagation for transitives is left to future work.
        scopes: directInfo ? directInfo.scopes : ["prod"],
        resolved: pkg.version
      };
      if (directInfo?.requested)
        entry.requested = directInfo.requested;
      if (sourceInfo.source)
        entry.source = sourceInfo.source;
      if (sourceInfo.sourceUrl)
        entry.sourceUrl = sourceInfo.sourceUrl;
      if (directInfo)
        directEntries.push(entry);
      else
        transitiveEntries.push(entry);
    }
    const manifest = {
      version: import_build_utils6.MANIFEST_VERSION,
      runtime: "rust",
      ...framework ? { framework } : {},
      ...serviceType ? { serviceType } : {},
      ...runtimeVersion ? { runtimeVersion } : {},
      dependencies: [
        ...directEntries.sort((a, b) => a.name.localeCompare(b.name)),
        ...transitiveEntries.sort((a, b) => a.name.localeCompare(b.name))
      ]
    };
    await (0, import_build_utils6.writeProjectManifest)(manifest, workPath, "rust");
  } catch {
  }
}
var diagnostics = (0, import_build_utils6.createDiagnostics)("rust");

// src/index.ts
async function buildHandler(options) {
  const BUILDER_DEBUG = Boolean(process.env.VERCEL_BUILDER_DEBUG ?? false);
  const isVercelBuild = Boolean(process.env.VERCEL_BUILD_IMAGE ?? false);
  const { files, entrypoint, workPath, config, meta, service } = options;
  const crossCompilationEnabled = !isVercelBuild && !meta?.isDev;
  if (crossCompilationEnabled && process.platform === "win32") {
    throw new Error(
      "Production prebuilt deployments for @vercel/rust are not yet supported on Windows. Please use a Linux or macOS environment, or deploy directly on Vercel."
    );
  }
  await installRustToolchain();
  (0, import_build_utils7.debug)("Creating file system");
  const downloadedFiles = await (0, import_build_utils7.download)(files, workPath, meta);
  const entryPath = downloadedFiles[entrypoint].fsPath;
  const HOME = process.platform === "win32" ? assertEnv("USERPROFILE") : assertEnv("HOME");
  const PATH = assertEnv("PATH");
  const rustEnv = {
    PATH: `${import_node_path4.default.join(HOME, ".cargo/bin")}${import_node_path4.default.delimiter}${PATH}`,
    RUSTFLAGS: [process.env.RUSTFLAGS].filter(Boolean).join(" ")
  };
  const cargoWorkspace = await findCargoWorkspace({
    env: rustEnv,
    cwd: import_node_path4.default.dirname(entryPath)
  });
  const binaryName = findBinaryName(cargoWorkspace, entryPath);
  const cargoBuildConfiguration = await findCargoBuildConfiguration(cargoWorkspace);
  await runUserScripts(workPath);
  const extraFiles = await gatherExtraFiles(config.includeFiles, workPath);
  const lambdaOptions = await (0, import_build_utils7.getLambdaOptionsFromFunction)({
    sourceFile: entrypoint,
    config
  });
  const architecture = lambdaOptions?.architecture || "x86_64";
  const buildVariant = meta?.isDev ? "debug" : "release";
  const buildTarget = cargoBuildConfiguration?.build.target ?? "";
  const targetTriple = architecture === "x86_64" ? "x86_64-unknown-linux-gnu" : "aarch64-unknown-linux-gnu";
  try {
    const args = crossCompilationEnabled ? ["zigbuild", "--target", targetTriple, "--bin", binaryName].concat(
      BUILDER_DEBUG ? ["--verbose"] : ["--quiet"],
      ["--release"]
    ) : ["build", "--bin", binaryName].concat(
      BUILDER_DEBUG ? ["--verbose"] : ["--quiet"],
      meta?.isDev ? [] : ["--release"]
    );
    (0, import_build_utils7.debug)(
      `Running \`cargo build\` for \`${binaryName}\` (\`${architecture}\`)`
    );
    await (0, import_execa4.default)("cargo", args, {
      cwd: workPath,
      env: rustEnv
    });
  } catch (err) {
    (0, import_build_utils7.debug)(`Running \`cargo build\` for \`${binaryName}\` failed`);
    throw err;
  }
  (0, import_build_utils7.debug)(
    `Building \`${binaryName}\` for \`${process.platform}\` (\`${architecture}\`) completed`
  );
  const cargoMetadata = await getCargoMetadata(
    { cwd: workPath, env: rustEnv },
    targetTriple
  );
  let { target_directory: targetDirectory } = cargoMetadata;
  if (crossCompilationEnabled) {
    targetDirectory = import_node_path4.default.join(targetDirectory, targetTriple);
  }
  targetDirectory = import_node_path4.default.join(targetDirectory, buildTarget);
  const bin = import_node_path4.default.join(
    targetDirectory,
    buildVariant,
    getExecutableName(binaryName)
  );
  const handler = getExecutableName("executable");
  const executableFile = new import_build_utils7.FileFsRef({ mode: 493, fsPath: bin });
  const lambda = new import_build_utils7.Lambda({
    ...lambdaOptions,
    files: {
      ...extraFiles,
      [handler]: executableFile
    },
    handler,
    supportsResponseStreaming: true,
    architecture,
    runtime: "executable",
    runtimeLanguage: "rust"
  });
  lambda.zipBuffer = await lambda.createZip();
  let resolvedRustVersion;
  try {
    const { stdout: rustcOut } = await (0, import_execa4.default)("rustc", ["--version"], {
      env: rustEnv,
      cwd: workPath
    });
    resolvedRustVersion = rustcOut.split(" ")[1];
  } catch {
    (0, import_build_utils7.debug)("Failed to determine rustc version");
  }
  const rootPkg = cargoMetadata.packages.find(
    (p) => p.id === cargoMetadata.resolve.root
  );
  const requestedRustVersion = rootPkg?.rust_version || void 0;
  await generateProjectManifest({
    workPath,
    cargoMetadata,
    framework: config?.framework ?? void 0,
    serviceType: service ? (0, import_build_utils7.getReportedServiceType)(service) : void 0,
    runtimeVersion: resolvedRustVersion ? {
      ...requestedRustVersion ? { requested: requestedRustVersion } : {},
      resolved: resolvedRustVersion
    } : void 0
  });
  (0, import_build_utils7.debug)(`generating function for \`${entrypoint}\``);
  return {
    output: lambda
  };
}
var runtime = {
  version: 3,
  build: buildHandler,
  prepareCache: async ({ workPath }) => {
    (0, import_build_utils7.debug)(`Caching \`${workPath}\``);
    const cacheFiles = await (0, import_build_utils7.glob)("target/**", workPath);
    for (const f of Object.keys(cacheFiles)) {
      const accept = /(?:^|\/)target\/release\/\.fingerprint\//.test(f) || /(?:^|\/)target\/release\/build\//.test(f) || /(?:^|\/)target\/release\/deps\//.test(f) || /(?:^|\/)target\/debug\/\.fingerprint\//.test(f) || /(?:^|\/)target\/debug\/build\//.test(f) || /(?:^|\/)target\/debug\/deps\//.test(f);
      if (!accept) {
        delete cacheFiles[f];
      }
    }
    return cacheFiles;
  },
  startDevServer,
  shouldServe: async (options) => {
    (0, import_build_utils7.debug)(`Requested ${options.requestPath} for ${options.entrypoint}`);
    const entrypointWithoutExt = options.entrypoint.replace(/\.rs$/, "");
    const matches = options.requestPath === options.entrypoint || options.requestPath === entrypointWithoutExt;
    (0, import_build_utils7.debug)(
      `shouldServe: ${matches} (entrypointWithoutExt: ${entrypointWithoutExt})`
    );
    return Promise.resolve(matches);
  }
};
var { version, build, prepareCache, startDevServer: startDevServer2, shouldServe } = runtime;
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  build,
  diagnostics,
  prepareCache,
  shouldServe,
  startDevServer,
  version
});
