// @vercel/sandbox/dist/api-client/api-client.cjs
const require_rolldown_runtime = require('../_virtual/rolldown_runtime.cjs');
const require_api_error = require('./api-error.cjs');
const require_base_client = require('./base-client.cjs');
const require_validators = require('./validators.cjs');
const require_file_writer = require('./file-writer.cjs');
const require_version = require('../version.cjs');
const require_consume_readable = require('../utils/consume-readable.cjs');
const require_normalizePath = require('../utils/normalizePath.cjs');
const require_network_policy = require('../utils/network-policy.cjs');
const require_types = require('../utils/types.cjs');
let zod = require("zod");
let stream = require("stream");
let jsonlines = require("jsonlines");
jsonlines = require_rolldown_runtime.__toESM(jsonlines);
let os = require("os");
os = require_rolldown_runtime.__toESM(os);
let __vercel_oidc = require("@vercel/oidc");

//#region src/api-client/api-client.ts
function decodeUnverifiedToken(token) {
	if (token.split(".").length !== 3) return null;
	try {
		const payload = JSON.parse(Buffer.from(token.split(".")[1], "base64url").toString("utf8"));
		if (payload.owner_id) return {
			owner_id: payload.owner_id,
			project_id: payload.project_id
		};
		return null;
	} catch {
		return null;
	}
}
var APIClient = class extends require_base_client.BaseClient {
	constructor(params) {
		super({
			baseUrl: params.baseUrl ?? "https://vercel.com/api",
			token: params.token,
			debug: false,
			fetch: params.fetch
		});
		this.teamId = params.teamId;
		this.isJwtToken = false;
		const claims = decodeUnverifiedToken(params.token);
		if (claims) {
			this.isJwtToken = true;
			this.projectId = claims.project_id;
			this.teamId = claims.owner_id;
		}
	}
	async ensureValidToken() {
		if (!this.isJwtToken) return;
		try {
			const freshToken = await (0, __vercel_oidc.getVercelOidcToken)({
				expirationBufferMs: 300 * 1e3,
				team: this.teamId,
				project: this.projectId
			});
			if (freshToken !== this.token) {
				this.token = freshToken;
				const claims = decodeUnverifiedToken(freshToken);
				if (claims) this.teamId = claims.owner_id;
			}
		} catch {}
	}
	async request(path, params) {
		await this.ensureValidToken();
		return super.request(path, {
			...params,
			query: {
				teamId: this.teamId,
				...params?.query
			},
			headers: {
				"content-type": "application/json",
				"user-agent": `vercel/sandbox/${require_version.VERSION} (Node.js/${process.version}; ${os.default.platform()}/${os.default.arch()})`,
				...params?.headers
			}
		});
	}
	async getSession(params) {
		const privateParams = require_types.getPrivateParams(params);
		let querystring = new URLSearchParams(privateParams).toString();
		querystring = querystring ? `?${querystring}` : "";
		return require_base_client.parseOrThrow(require_validators.SessionAndRoutesResponse, await this.request(`/v2/sandboxes/sessions/${params.sessionId}${querystring}`, { signal: params.signal }));
	}
	async createSandbox(params) {
		const privateParams = require_types.getPrivateParams(params);
		return require_base_client.parseOrThrow(require_validators.SandboxAndSessionResponse, await this.request("/v2/sandboxes", {
			method: "POST",
			body: JSON.stringify({
				projectId: params.projectId,
				ports: params.ports,
				source: params.source,
				timeout: params.timeout,
				resources: params.resources,
				runtime: params.runtime,
				image: params.image,
				name: params.name,
				persistent: params.persistent,
				networkPolicy: params.networkPolicy ? require_network_policy.toAPINetworkPolicy(params.networkPolicy) : void 0,
				env: params.env,
				tags: params.tags,
				snapshotExpiration: params.snapshotExpiration,
				keepLastSnapshots: params.keepLastSnapshots,
				...privateParams
			}),
			signal: params.signal
		}));
	}
	async runCommand(params) {
		if (params.wait) {
			const response = await this.request(`/v2/sandboxes/sessions/${params.sessionId}/cmd`, {
				method: "POST",
				body: JSON.stringify({
					command: params.command,
					args: params.args,
					cwd: params.cwd,
					env: params.env,
					sudo: params.sudo,
					wait: true,
					logs: params.logs || void 0,
					timeout: params.timeout
				}),
				signal: params.signal
			});
			if (!response.ok) await require_base_client.parseOrThrow(zod.z.any(), response);
			if (response.headers.get("content-type") !== "application/x-ndjson") throw new require_api_error.APIError(response, {
				message: "Expected a stream of command data",
				sessionId: params.sessionId
			});
			if (response.body === null) throw new require_api_error.APIError(response, {
				message: "No response body",
				sessionId: params.sessionId
			});
			const jsonlinesStream = jsonlines.default.parse();
			pipe(response.body, jsonlinesStream, { signal: params.signal }).catch((err) => {
				console.error("Error piping command stream:", err);
			});
			const iterator = jsonlinesStream[Symbol.asyncIterator]();
			const commandChunk = await iterator.next();
			if (commandChunk.done) throw new require_api_error.StreamError("stream_ended_early", "Stream ended before command data was received", params.sessionId);
			const { command } = require_validators.CommandResponse.parse(commandChunk.value);
			return {
				command,
				finished: (async () => {
					while (true) {
						const chunk = await iterator.next();
						if (chunk.done) throw new require_api_error.StreamError("stream_ended_early", "Stream ended before command finished", params.sessionId);
						if (chunk.value?.command) {
							const { command: command$1 } = require_validators.CommandFinishedResponse.parse(chunk.value);
							return command$1;
						}
						const parsed = require_validators.LogLine.parse(chunk.value);
						if (parsed.stream === "error") throw new require_api_error.StreamError(parsed.data.code, parsed.data.message, params.sessionId);
						params.onLog?.(parsed);
					}
				})()
			};
		}
		return require_base_client.parseOrThrow(require_validators.CommandResponse, await this.request(`/v2/sandboxes/sessions/${params.sessionId}/cmd`, {
			method: "POST",
			body: JSON.stringify({
				command: params.command,
				args: params.args,
				cwd: params.cwd,
				env: params.env,
				sudo: params.sudo,
				timeout: params.timeout
			}),
			signal: params.signal
		}));
	}
	async getCommand(params) {
		return params.wait ? require_base_client.parseOrThrow(require_validators.CommandFinishedResponse, await this.request(`/v2/sandboxes/sessions/${params.sessionId}/cmd/${params.cmdId}`, {
			signal: params.signal,
			query: { wait: "true" }
		})) : require_base_client.parseOrThrow(require_validators.CommandResponse, await this.request(`/v2/sandboxes/sessions/${params.sessionId}/cmd/${params.cmdId}`, { signal: params.signal }));
	}
	async openInteractive(params) {
		return require_base_client.parseOrThrow(require_validators.InteractiveSessionResponse, await this.request(`/v2/sandboxes/sessions/${params.sessionId}/interactive`, {
			method: "POST",
			body: JSON.stringify({}),
			signal: params.signal
		}));
	}
	async mkDir(params) {
		return require_base_client.parseOrThrow(require_validators.EmptyResponse, await this.request(`/v2/sandboxes/sessions/${params.sessionId}/fs/mkdir`, {
			method: "POST",
			body: JSON.stringify({
				path: params.path,
				cwd: params.cwd
			}),
			signal: params.signal
		}));
	}
	getFileWriter(params) {
		const writer = new require_file_writer.FileWriter();
		return {
			response: (async () => {
				return this.request(`/v2/sandboxes/sessions/${params.sessionId}/fs/write`, {
					method: "POST",
					headers: {
						"content-type": "application/gzip",
						"x-cwd": params.extractDir
					},
					body: await require_consume_readable.consumeReadable(writer.readable),
					signal: params.signal
				});
			})(),
			writer
		};
	}
	async listSessions(params) {
		return require_base_client.parseOrThrow(require_validators.SessionsResponse, await this.request(`/v2/sandboxes/sessions`, {
			query: {
				project: params.projectId,
				name: params.name,
				limit: params.limit,
				cursor: params.cursor,
				sortOrder: params.sortOrder
			},
			method: "GET",
			signal: params.signal
		}));
	}
	async listSnapshots(params) {
		return require_base_client.parseOrThrow(require_validators.SnapshotsResponse, await this.request(`/v2/sandboxes/snapshots`, {
			query: {
				project: params.projectId,
				name: params.name,
				limit: params.limit,
				cursor: params.cursor,
				sortOrder: params.sortOrder
			},
			method: "GET",
			signal: params.signal
		}));
	}
	async getSnapshotTree(params) {
		return require_base_client.parseOrThrow(require_validators.SnapshotTreeResponse, await this.request(`/v2/sandboxes/snapshots/tree`, {
			query: {
				project: params.projectId,
				snapshotId: params.snapshotId,
				limit: params.limit,
				sortOrder: params.sortOrder
			},
			method: "GET",
			signal: params.signal
		}));
	}
	async writeFiles(params) {
		const { writer, response } = this.getFileWriter({
			sessionId: params.sessionId,
			extractDir: params.extractDir,
			signal: params.signal
		});
		for (const file of params.files) await writer.addFile({
			name: require_normalizePath.normalizePath({
				filePath: file.path,
				extractDir: params.extractDir,
				cwd: params.cwd
			}),
			content: file.content,
			mode: file.mode
		});
		writer.end();
		await require_base_client.parseOrThrow(require_validators.EmptyResponse, await response);
	}
	async readFile(params) {
		const response = await this.request(`/v2/sandboxes/sessions/${params.sessionId}/fs/read`, {
			method: "POST",
			body: JSON.stringify({
				path: params.path,
				cwd: params.cwd
			}),
			signal: params.signal
		});
		if (response.status === 404) return null;
		const contentType = response.headers.get("content-type") ?? "";
		if (!response.ok || !contentType.includes("application/octet-stream")) {
			const text = await response.text().catch(() => "");
			let json;
			try {
				json = JSON.parse(text || "{}");
			} catch {
				json = void 0;
			}
			throw new require_api_error.APIError(response, {
				message: `Unexpected response reading file: status ${response.status}, content-type ${contentType || "(none)"}`,
				json,
				text
			});
		}
		if (response.body === null) return null;
		return stream.Readable.fromWeb(response.body);
	}
	async killCommand(params) {
		return require_base_client.parseOrThrow(require_validators.CommandResponse, await this.request(`/v2/sandboxes/sessions/${params.sessionId}/cmd/${params.commandId}/kill`, {
			method: "POST",
			body: JSON.stringify({ signal: params.signal }),
			signal: params.abortSignal
		}));
	}
	getLogs(params) {
		const self = this;
		const disposer = new AbortController();
		const signal = !params.signal ? disposer.signal : mergeSignals(params.signal, disposer.signal);
		const generator = (async function* () {
			const url = `/v2/sandboxes/sessions/${params.sessionId}/cmd/${params.cmdId}/logs`;
			const response = await self.request(url, {
				method: "GET",
				signal
			});
			if (!response.ok) await require_base_client.parseOrThrow(zod.z.any(), response);
			if (response.headers.get("content-type") !== "application/x-ndjson") throw new require_api_error.APIError(response, {
				message: "Expected a stream of logs",
				sessionId: params.sessionId
			});
			if (response.body === null) throw new require_api_error.APIError(response, {
				message: "No response body",
				sessionId: params.sessionId
			});
			const jsonlinesStream = jsonlines.default.parse();
			pipe(response.body, jsonlinesStream, { signal }).catch((err) => {
				console.error("Error piping logs:", err);
			});
			for await (const chunk of jsonlinesStream) {
				const parsed = require_validators.LogLine.parse(chunk);
				if (parsed.stream === "error") throw new require_api_error.StreamError(parsed.data.code, parsed.data.message, params.sessionId);
				yield parsed;
			}
		})();
		return Object.assign(generator, {
			[Symbol.dispose]() {
				disposer.abort("Disposed");
			},
			close: () => disposer.abort("Disposed")
		});
	}
	async stopSession(params) {
		const url = `/v2/sandboxes/sessions/${params.sessionId}/stop`;
		return require_base_client.parseOrThrow(require_validators.StopSessionResponse, await this.request(url, {
			method: "POST",
			signal: params.signal
		}));
	}
	async updateNetworkPolicy(params) {
		const url = `/v2/sandboxes/sessions/${params.sessionId}/network-policy`;
		return require_base_client.parseOrThrow(require_validators.SessionResponse, await this.request(url, {
			method: "POST",
			body: JSON.stringify(require_network_policy.toAPINetworkPolicy(params.networkPolicy)),
			signal: params.signal
		}));
	}
	async extendTimeout(params) {
		const url = `/v2/sandboxes/sessions/${params.sessionId}/extend-timeout`;
		return require_base_client.parseOrThrow(require_validators.SessionResponse, await this.request(url, {
			method: "POST",
			body: JSON.stringify({ duration: params.duration }),
			signal: params.signal
		}));
	}
	async createSnapshot(params) {
		const url = `/v2/sandboxes/sessions/${params.sessionId}/snapshot`;
		const body = params.expiration === void 0 ? void 0 : JSON.stringify({ expiration: params.expiration });
		return require_base_client.parseOrThrow(require_validators.CreateSnapshotResponse, await this.request(url, {
			method: "POST",
			body,
			signal: params.signal
		}));
	}
	async deleteSnapshot(params) {
		const url = `/v2/sandboxes/snapshots/${params.snapshotId}`;
		return require_base_client.parseOrThrow(require_validators.SnapshotResponse, await this.request(url, {
			method: "DELETE",
			signal: params.signal
		}));
	}
	async getSnapshot(params) {
		const url = `/v2/sandboxes/snapshots/${params.snapshotId}`;
		return require_base_client.parseOrThrow(require_validators.SnapshotResponse, await this.request(url, { signal: params.signal }));
	}
	async getSandbox(params) {
		const privateParams = require_types.getPrivateParams(params);
		const query = {
			projectId: params.projectId,
			...privateParams
		};
		if (params.resume !== void 0) query.resume = String(params.resume);
		return require_base_client.parseOrThrow(require_validators.SandboxAndSessionResponse, await this.request(`/v2/sandboxes/${encodeURIComponent(params.name)}`, {
			query,
			signal: params.signal
		}));
	}
	async listSandboxes(params) {
		return require_base_client.parseOrThrow(require_validators.SandboxesPaginationResponse, await this.request(`/v2/sandboxes`, {
			query: {
				project: params.projectId,
				limit: params.limit,
				sortBy: params.sortBy,
				sortOrder: params.sortOrder,
				namePrefix: params.namePrefix,
				cursor: params.cursor,
				tags: toTagsFilter(params.tags)
			},
			method: "GET",
			signal: params.signal
		}));
	}
	async updateSandbox(params) {
		return require_base_client.parseOrThrow(require_validators.UpdateSandboxResponse, await this.request(`/v2/sandboxes/${encodeURIComponent(params.name)}`, {
			method: "PATCH",
			query: { projectId: params.projectId },
			body: JSON.stringify({
				persistent: params.persistent,
				resources: params.resources,
				runtime: params.runtime,
				timeout: params.timeout,
				networkPolicy: params.networkPolicy ? require_network_policy.toAPINetworkPolicy(params.networkPolicy) : void 0,
				tags: params.tags,
				ports: params.ports,
				snapshotExpiration: params.snapshotExpiration,
				keepLastSnapshots: params.keepLastSnapshots,
				currentSnapshotId: params.currentSnapshotId
			}),
			signal: params.signal
		}));
	}
	async deleteSandbox(params) {
		return require_base_client.parseOrThrow(require_validators.UpdateSandboxResponse, await this.request(`/v2/sandboxes/${encodeURIComponent(params.name)}`, {
			method: "DELETE",
			query: { projectId: params.projectId },
			signal: params.signal
		}));
	}
};
async function pipe(readable, output, options) {
	const reader = readable.getReader();
	let aborted = false;
	const signal = options?.signal;
	const onAbort = () => {
		aborted = true;
		const reason = signal?.reason ?? new DOMException("The operation was aborted.", "AbortError");
		reader.cancel(reason).catch(() => {});
		if ("destroy" in output && typeof output.destroy === "function") {
			output.destroy(reason);
			return;
		}
		output.emit("error", reason);
		output.end();
	};
	if (signal) if (signal.aborted) onAbort();
	else signal.addEventListener("abort", onAbort, { once: true });
	try {
		while (true) {
			const read = await reader.read();
			if (read.value) output.write(Buffer.from(read.value));
			if (read.done) break;
		}
	} catch (err) {
		if (!aborted) output.emit("error", err);
	} finally {
		signal?.removeEventListener("abort", onAbort);
		if (!aborted) output.end();
	}
}
function mergeSignals(...signals) {
	const controller = new AbortController();
	const onAbort = () => {
		controller.abort();
		for (const signal of signals) signal.removeEventListener("abort", onAbort);
	};
	for (const signal of signals) {
		if (signal.aborted) {
			controller.abort();
			break;
		}
		signal.addEventListener("abort", onAbort);
	}
	return controller.signal;
}
function toTagsFilter(tags) {
	if (tags === void 0) return void 0;
	const entries = Object.entries(tags);
	if (entries.length === 0) return void 0;
	return entries.map(([key, value]) => `${key}:${value}`);
}

//#endregion
exports.APIClient = APIClient;
