// @vercel/sandbox/dist/api-client/base-client.js
import { APIError } from "./api-error.js";
import { array } from "../utils/array.js";
import { withRetry } from "./with-retry.js";
import { Agent } from "undici";

//#region src/api-client/base-client.ts
const DEFAULT_AGENT = new Agent({ bodyTimeout: 0 });
/**
* A base API client that provides a convenience wrapper for fetching where
* we can pass query parameters as an object, support retries, debugging
* and automatic authorization.
*/
var BaseClient = class {
	constructor(params) {
		this.fetch = withRetry(params.fetch ?? globalThis.fetch);
		this.baseUrl = params.baseUrl;
		this.debug = params.debug ?? process.env.DEBUG_FETCH === "true";
		this.token = params.token;
		this.agent = DEFAULT_AGENT;
	}
	async request(path, opts) {
		const url = new URL(`${this.baseUrl}${path}`);
		if (opts?.query) for (const [key, value] of Object.entries(opts.query)) array(value).forEach((value$1) => {
			url.searchParams.append(key, value$1.toString());
		});
		const start = Date.now();
		const response = await this.fetch(url.toString(), {
			...opts,
			body: opts?.body,
			method: opts?.method || "GET",
			headers: this.token ? {
				Authorization: `Bearer ${this.token}`,
				...opts?.headers
			} : opts?.headers,
			dispatcher: this.agent,
			signal: opts?.signal
		});
		if (this.debug) {
			const duration = Date.now() - start;
			console.log(`[API] ${url} (${response.status}) ${duration}ms`);
			if (response.status === 429) {
				const retry = parseInt(response.headers.get("Retry-After") ?? "", 10);
				const hours = Math.floor(retry / 60 / 60);
				const minutes = Math.floor(retry / 60) % 60;
				const seconds = retry % 60;
				console.warn(`[API] ${url} Rate Limited, Retry After ${hours}h ${minutes}m ${seconds}s`);
			}
		}
		return response;
	}
};
/**
* Extract sessionId from a sandbox API URL.
*/
function extractSessionId(url) {
	return url.match(/\/v2\/sandboxes\/sessions\/([^/?]+)/)?.[1];
}
/**
* Extract sandbox name from a sandbox API url.
* Excludes known sub-paths like /sessions/ and /snapshots/.
*/
function extractSandboxName(url) {
	return url.match(/\/v2\/sandboxes\/(?!sessions(?:\/|$|\?))(?!snapshots(?:\/|$|\?))([^/?]+)/)?.[1];
}
/**
* Allows to read the response text and parse it as JSON casting to the given
* type. If the response is not ok or cannot be parsed it will return error.
*
* @param response Response to parse.
* @returns Parsed response or error.
*/
async function parse(validator, response) {
	const sessionId = extractSessionId(response.url);
	let sandboxName;
	if (!sessionId) sandboxName = extractSandboxName(response.url);
	const text = await response.text().catch((err) => {
		return new APIError(response, {
			message: `Can't read response text: ${String(err)}`,
			sessionId,
			sandboxName
		});
	});
	if (typeof text !== "string") return text;
	let json;
	try {
		json = JSON.parse(text || "{}");
	} catch (error) {
		return new APIError(response, {
			message: `Can't parse JSON: ${String(error)}`,
			text,
			sessionId,
			sandboxName
		});
	}
	if (!response.ok) return new APIError(response, {
		message: `Status code ${response.status} is not ok`,
		json,
		text,
		sessionId,
		sandboxName
	});
	const validated = validator.safeParse(json);
	if (!validated.success) return new APIError(response, {
		message: `Response JSON is not valid: ${validated.error}`,
		json,
		text,
		sessionId,
		sandboxName
	});
	return {
		json: validated.data,
		response,
		text
	};
}
async function parseOrThrow(validator, response) {
	const result = await parse(validator, response);
	if (result instanceof APIError) throw result;
	return result;
}

//#endregion
export { BaseClient, parseOrThrow };
