// @vercel/sandbox/dist/api-client/validators.cjs
const require_rolldown_runtime = require('../_virtual/rolldown_runtime.cjs');
let zod = require("zod");

//#region src/api-client/validators.ts
const RuleMatcherValidator = zod.z.object({
	exact: zod.z.string().optional(),
	startsWith: zod.z.string().optional(),
	regex: zod.z.string().optional()
});
const KeyValueMatcherValidator = zod.z.object({
	key: RuleMatcherValidator.optional(),
	value: RuleMatcherValidator.optional()
});
const RuleMatchValidator = zod.z.object({
	path: RuleMatcherValidator.optional(),
	method: zod.z.array(zod.z.string()).optional(),
	queryString: zod.z.array(KeyValueMatcherValidator).optional(),
	headers: zod.z.array(KeyValueMatcherValidator).optional()
});
const InjectionRuleValidator = zod.z.object({
	domain: zod.z.string(),
	headers: zod.z.record(zod.z.string(), zod.z.string()).optional(),
	headerNames: zod.z.array(zod.z.string()).optional(),
	match: RuleMatchValidator.optional()
});
const ForwardRuleValidator = zod.z.object({
	domain: zod.z.string(),
	forwardURL: zod.z.string(),
	match: RuleMatchValidator.optional()
});
const NetworkPolicyTransformValidator = zod.z.object({ headers: zod.z.record(zod.z.string(), zod.z.string()).optional() });
const NetworkPolicyRuleValidator = zod.z.object({
	match: RuleMatchValidator.optional(),
	transform: zod.z.array(NetworkPolicyTransformValidator).optional(),
	forwardURL: zod.z.string().optional()
});
const V2NetworkPolicyObjectValidator = zod.z.object({
	allow: zod.z.union([zod.z.array(zod.z.string()), zod.z.record(zod.z.string(), zod.z.array(NetworkPolicyRuleValidator))]).optional(),
	subnets: zod.z.object({
		allow: zod.z.array(zod.z.string()).optional(),
		deny: zod.z.array(zod.z.string()).optional()
	}).optional()
});
const NetworkPolicyModeValidator = zod.z.union([zod.z.object({ mode: zod.z.literal("allow-all") }).passthrough(), zod.z.object({ mode: zod.z.literal("deny-all") }).passthrough()]);
const LegacyCustomNetworkPolicyValidator = zod.z.object({
	mode: zod.z.literal("custom"),
	allowedDomains: zod.z.array(zod.z.string()).optional(),
	allowedCIDRs: zod.z.array(zod.z.string()).optional(),
	deniedCIDRs: zod.z.array(zod.z.string()).optional(),
	injectionRules: zod.z.array(InjectionRuleValidator).optional(),
	forwardRules: zod.z.array(ForwardRuleValidator).optional()
}).passthrough();
const NetworkPolicyRequestValidator = zod.z.union([NetworkPolicyModeValidator, V2NetworkPolicyObjectValidator.passthrough()]);
const NetworkPolicyResponseValidator = zod.z.union([NetworkPolicyModeValidator, LegacyCustomNetworkPolicyValidator]);
const Session = zod.z.object({
	id: zod.z.string(),
	memory: zod.z.number(),
	vcpus: zod.z.number(),
	region: zod.z.string(),
	runtime: zod.z.string(),
	timeout: zod.z.number(),
	status: zod.z.enum([
		"pending",
		"running",
		"stopping",
		"stopped",
		"failed",
		"aborted",
		"snapshotting"
	]),
	requestedAt: zod.z.number(),
	startedAt: zod.z.number().optional(),
	requestedStopAt: zod.z.number().optional(),
	stoppedAt: zod.z.number().optional(),
	abortedAt: zod.z.number().optional(),
	duration: zod.z.number().optional(),
	sourceSnapshotId: zod.z.string().optional(),
	snapshottedAt: zod.z.number().optional(),
	createdAt: zod.z.number(),
	cwd: zod.z.string(),
	updatedAt: zod.z.number(),
	interactivePort: zod.z.number().optional(),
	networkPolicy: NetworkPolicyResponseValidator.optional(),
	activeCpuDurationMs: zod.z.number().optional(),
	networkTransfer: zod.z.object({
		ingress: zod.z.number(),
		egress: zod.z.number()
	}).optional()
});
const SandboxRoute = zod.z.object({
	url: zod.z.string(),
	subdomain: zod.z.string(),
	port: zod.z.number()
});
const Snapshot = zod.z.object({
	id: zod.z.string(),
	sourceSessionId: zod.z.string(),
	region: zod.z.string(),
	status: zod.z.enum([
		"created",
		"deleted",
		"failed"
	]),
	sizeBytes: zod.z.number(),
	expiresAt: zod.z.number().optional(),
	createdAt: zod.z.number(),
	updatedAt: zod.z.number(),
	lastUsedAt: zod.z.number().optional(),
	creationMethod: zod.z.string().optional(),
	parentId: zod.z.string().optional()
});
const CursorPagination = zod.z.object({
	count: zod.z.number(),
	next: zod.z.string().nullable()
});
const Command = zod.z.object({
	id: zod.z.string(),
	name: zod.z.string(),
	args: zod.z.array(zod.z.string()),
	cwd: zod.z.string(),
	sessionId: zod.z.string(),
	exitCode: zod.z.number().nullable(),
	startedAt: zod.z.number()
});
const CommandFinished = Command.extend({ exitCode: zod.z.number() });
const SessionResponse = zod.z.object({ session: Session.passthrough() });
const SessionAndRoutesResponse = SessionResponse.extend({ routes: zod.z.array(SandboxRoute) });
const InteractiveSessionResponse = zod.z.object({
	url: zod.z.string(),
	token: zod.z.string()
});
const SessionsResponse = zod.z.object({
	sessions: zod.z.array(Session.passthrough()),
	pagination: CursorPagination
});
const CommandResponse = zod.z.object({ command: Command });
const CommandFinishedResponse = zod.z.object({ command: CommandFinished });
const EmptyResponse = zod.z.object({});
const LogLineBase = zod.z.object({ data: zod.z.string() });
const LogLineStdout = LogLineBase.extend({ stream: zod.z.literal("stdout") });
const LogLineStderr = LogLineBase.extend({ stream: zod.z.literal("stderr") });
const LogError = zod.z.object({
	stream: zod.z.literal("error"),
	data: zod.z.object({
		code: zod.z.string(),
		message: zod.z.string()
	})
});
const LogLine = zod.z.discriminatedUnion("stream", [
	LogLineStdout,
	LogLineStderr,
	LogError
]);
const SnapshotsResponse = zod.z.object({
	snapshots: zod.z.array(Snapshot),
	pagination: CursorPagination
});
const SnapshotTreeNode = zod.z.object({
	snapshot: Snapshot,
	siblings: zod.z.array(Snapshot),
	count: zod.z.string()
});
const SnapshotTreeResponse = zod.z.object({
	snapshots: zod.z.array(SnapshotTreeNode),
	anchor: SnapshotTreeNode.optional(),
	pagination: CursorPagination
});
const CreateSnapshotResponse = zod.z.object({
	snapshot: Snapshot,
	session: Session.passthrough()
});
const SnapshotResponse = zod.z.object({ snapshot: Snapshot });
const Sandbox = zod.z.object({
	name: zod.z.string(),
	persistent: zod.z.boolean(),
	region: zod.z.string().optional(),
	vcpus: zod.z.number().optional(),
	memory: zod.z.number().optional(),
	runtime: zod.z.string().optional(),
	timeout: zod.z.number().optional(),
	networkPolicy: NetworkPolicyResponseValidator.optional(),
	totalEgressBytes: zod.z.number().optional(),
	totalIngressBytes: zod.z.number().optional(),
	totalActiveCpuDurationMs: zod.z.number().optional(),
	totalDurationMs: zod.z.number().optional(),
	createdAt: zod.z.number(),
	updatedAt: zod.z.number(),
	expiresAt: zod.z.number().optional(),
	currentSessionId: zod.z.string(),
	currentSnapshotId: zod.z.string().optional(),
	status: Session.shape.status,
	statusUpdatedAt: zod.z.number().optional(),
	cwd: zod.z.string().optional(),
	tags: zod.z.record(zod.z.string(), zod.z.string()).optional(),
	snapshotExpiration: zod.z.number().optional(),
	keepLastSnapshots: zod.z.object({
		count: zod.z.number(),
		expiration: zod.z.number().optional(),
		deleteEvicted: zod.z.boolean().optional()
	}).optional()
});
const StopSessionResponse = zod.z.object({
	session: Session.passthrough(),
	sandbox: Sandbox.optional(),
	snapshot: Snapshot.optional()
});
const SandboxAndSessionResponse = zod.z.object({
	sandbox: Sandbox,
	session: Session.passthrough(),
	routes: zod.z.array(SandboxRoute),
	resumed: zod.z.boolean().optional()
});
const SandboxesPaginationResponse = zod.z.object({
	sandboxes: zod.z.array(Sandbox),
	pagination: CursorPagination
});
const UpdateSandboxResponse = zod.z.object({
	sandbox: Sandbox,
	routes: zod.z.array(SandboxRoute).optional()
});

//#endregion
exports.Command = Command;
exports.CommandFinishedResponse = CommandFinishedResponse;
exports.CommandResponse = CommandResponse;
exports.CreateSnapshotResponse = CreateSnapshotResponse;
exports.CursorPagination = CursorPagination;
exports.EmptyResponse = EmptyResponse;
exports.ForwardRuleValidator = ForwardRuleValidator;
exports.InjectionRuleValidator = InjectionRuleValidator;
exports.InteractiveSessionResponse = InteractiveSessionResponse;
exports.LogError = LogError;
exports.LogLine = LogLine;
exports.LogLineStderr = LogLineStderr;
exports.LogLineStdout = LogLineStdout;
exports.NetworkPolicyResponseValidator = NetworkPolicyResponseValidator;
exports.NetworkPolicyRuleValidator = NetworkPolicyRuleValidator;
exports.NetworkPolicyTransformValidator = NetworkPolicyTransformValidator;
exports.Sandbox = Sandbox;
exports.SandboxAndSessionResponse = SandboxAndSessionResponse;
exports.SandboxRoute = SandboxRoute;
exports.SandboxesPaginationResponse = SandboxesPaginationResponse;
exports.Session = Session;
exports.SessionAndRoutesResponse = SessionAndRoutesResponse;
exports.SessionResponse = SessionResponse;
exports.SessionsResponse = SessionsResponse;
exports.Snapshot = Snapshot;
exports.SnapshotResponse = SnapshotResponse;
exports.SnapshotTreeNode = SnapshotTreeNode;
exports.SnapshotTreeResponse = SnapshotTreeResponse;
exports.SnapshotsResponse = SnapshotsResponse;
exports.StopSessionResponse = StopSessionResponse;
exports.UpdateSandboxResponse = UpdateSandboxResponse;
exports.V2NetworkPolicyObjectValidator = V2NetworkPolicyObjectValidator;
