// @vercel/sandbox/dist/api-client/validators.js
import { z } from "zod";

//#region src/api-client/validators.ts
const RuleMatcherValidator = z.object({
	exact: z.string().optional(),
	startsWith: z.string().optional(),
	regex: z.string().optional()
});
const KeyValueMatcherValidator = z.object({
	key: RuleMatcherValidator.optional(),
	value: RuleMatcherValidator.optional()
});
const RuleMatchValidator = z.object({
	path: RuleMatcherValidator.optional(),
	method: z.array(z.string()).optional(),
	queryString: z.array(KeyValueMatcherValidator).optional(),
	headers: z.array(KeyValueMatcherValidator).optional()
});
const InjectionRuleValidator = z.object({
	domain: z.string(),
	headers: z.record(z.string(), z.string()).optional(),
	headerNames: z.array(z.string()).optional(),
	match: RuleMatchValidator.optional()
});
const ForwardRuleValidator = z.object({
	domain: z.string(),
	forwardURL: z.string(),
	match: RuleMatchValidator.optional()
});
const NetworkPolicyTransformValidator = z.object({ headers: z.record(z.string(), z.string()).optional() });
const NetworkPolicyRuleValidator = z.object({
	match: RuleMatchValidator.optional(),
	transform: z.array(NetworkPolicyTransformValidator).optional(),
	forwardURL: z.string().optional()
});
const V2NetworkPolicyObjectValidator = z.object({
	allow: z.union([z.array(z.string()), z.record(z.string(), z.array(NetworkPolicyRuleValidator))]).optional(),
	subnets: z.object({
		allow: z.array(z.string()).optional(),
		deny: z.array(z.string()).optional()
	}).optional()
});
const NetworkPolicyModeValidator = z.union([z.object({ mode: z.literal("allow-all") }).passthrough(), z.object({ mode: z.literal("deny-all") }).passthrough()]);
const LegacyCustomNetworkPolicyValidator = z.object({
	mode: z.literal("custom"),
	allowedDomains: z.array(z.string()).optional(),
	allowedCIDRs: z.array(z.string()).optional(),
	deniedCIDRs: z.array(z.string()).optional(),
	injectionRules: z.array(InjectionRuleValidator).optional(),
	forwardRules: z.array(ForwardRuleValidator).optional()
}).passthrough();
const NetworkPolicyRequestValidator = z.union([NetworkPolicyModeValidator, V2NetworkPolicyObjectValidator.passthrough()]);
const NetworkPolicyResponseValidator = z.union([NetworkPolicyModeValidator, LegacyCustomNetworkPolicyValidator]);
const Session = z.object({
	id: z.string(),
	memory: z.number(),
	vcpus: z.number(),
	region: z.string(),
	runtime: z.string(),
	timeout: z.number(),
	status: z.enum([
		"pending",
		"running",
		"stopping",
		"stopped",
		"failed",
		"aborted",
		"snapshotting"
	]),
	requestedAt: z.number(),
	startedAt: z.number().optional(),
	requestedStopAt: z.number().optional(),
	stoppedAt: z.number().optional(),
	abortedAt: z.number().optional(),
	duration: z.number().optional(),
	sourceSnapshotId: z.string().optional(),
	snapshottedAt: z.number().optional(),
	createdAt: z.number(),
	cwd: z.string(),
	updatedAt: z.number(),
	interactivePort: z.number().optional(),
	networkPolicy: NetworkPolicyResponseValidator.optional(),
	activeCpuDurationMs: z.number().optional(),
	networkTransfer: z.object({
		ingress: z.number(),
		egress: z.number()
	}).optional()
});
const SandboxRoute = z.object({
	url: z.string(),
	subdomain: z.string(),
	port: z.number()
});
const Snapshot = z.object({
	id: z.string(),
	sourceSessionId: z.string(),
	region: z.string(),
	status: z.enum([
		"created",
		"deleted",
		"failed"
	]),
	sizeBytes: z.number(),
	expiresAt: z.number().optional(),
	createdAt: z.number(),
	updatedAt: z.number(),
	lastUsedAt: z.number().optional(),
	creationMethod: z.string().optional(),
	parentId: z.string().optional()
});
const CursorPagination = z.object({
	count: z.number(),
	next: z.string().nullable()
});
const Command = z.object({
	id: z.string(),
	name: z.string(),
	args: z.array(z.string()),
	cwd: z.string(),
	sessionId: z.string(),
	exitCode: z.number().nullable(),
	startedAt: z.number()
});
const CommandFinished = Command.extend({ exitCode: z.number() });
const SessionResponse = z.object({ session: Session.passthrough() });
const SessionAndRoutesResponse = SessionResponse.extend({ routes: z.array(SandboxRoute) });
const InteractiveSessionResponse = z.object({
	url: z.string(),
	token: z.string()
});
const SessionsResponse = z.object({
	sessions: z.array(Session.passthrough()),
	pagination: CursorPagination
});
const CommandResponse = z.object({ command: Command });
const CommandFinishedResponse = z.object({ command: CommandFinished });
const EmptyResponse = z.object({});
const LogLineBase = z.object({ data: z.string() });
const LogLineStdout = LogLineBase.extend({ stream: z.literal("stdout") });
const LogLineStderr = LogLineBase.extend({ stream: z.literal("stderr") });
const LogError = z.object({
	stream: z.literal("error"),
	data: z.object({
		code: z.string(),
		message: z.string()
	})
});
const LogLine = z.discriminatedUnion("stream", [
	LogLineStdout,
	LogLineStderr,
	LogError
]);
const SnapshotsResponse = z.object({
	snapshots: z.array(Snapshot),
	pagination: CursorPagination
});
const SnapshotTreeNode = z.object({
	snapshot: Snapshot,
	siblings: z.array(Snapshot),
	count: z.string()
});
const SnapshotTreeResponse = z.object({
	snapshots: z.array(SnapshotTreeNode),
	anchor: SnapshotTreeNode.optional(),
	pagination: CursorPagination
});
const CreateSnapshotResponse = z.object({
	snapshot: Snapshot,
	session: Session.passthrough()
});
const SnapshotResponse = z.object({ snapshot: Snapshot });
const Sandbox = z.object({
	name: z.string(),
	persistent: z.boolean(),
	region: z.string().optional(),
	vcpus: z.number().optional(),
	memory: z.number().optional(),
	runtime: z.string().optional(),
	timeout: z.number().optional(),
	networkPolicy: NetworkPolicyResponseValidator.optional(),
	totalEgressBytes: z.number().optional(),
	totalIngressBytes: z.number().optional(),
	totalActiveCpuDurationMs: z.number().optional(),
	totalDurationMs: z.number().optional(),
	createdAt: z.number(),
	updatedAt: z.number(),
	expiresAt: z.number().optional(),
	currentSessionId: z.string(),
	currentSnapshotId: z.string().optional(),
	status: Session.shape.status,
	statusUpdatedAt: z.number().optional(),
	cwd: z.string().optional(),
	tags: z.record(z.string(), z.string()).optional(),
	snapshotExpiration: z.number().optional(),
	keepLastSnapshots: z.object({
		count: z.number(),
		expiration: z.number().optional(),
		deleteEvicted: z.boolean().optional()
	}).optional()
});
const StopSessionResponse = z.object({
	session: Session.passthrough(),
	sandbox: Sandbox.optional(),
	snapshot: Snapshot.optional()
});
const SandboxAndSessionResponse = z.object({
	sandbox: Sandbox,
	session: Session.passthrough(),
	routes: z.array(SandboxRoute),
	resumed: z.boolean().optional()
});
const SandboxesPaginationResponse = z.object({
	sandboxes: z.array(Sandbox),
	pagination: CursorPagination
});
const UpdateSandboxResponse = z.object({
	sandbox: Sandbox,
	routes: z.array(SandboxRoute).optional()
});

//#endregion
export { Command, CommandFinishedResponse, CommandResponse, CreateSnapshotResponse, CursorPagination, EmptyResponse, ForwardRuleValidator, InjectionRuleValidator, InteractiveSessionResponse, LogError, LogLine, LogLineStderr, LogLineStdout, NetworkPolicyResponseValidator, NetworkPolicyRuleValidator, NetworkPolicyTransformValidator, Sandbox, SandboxAndSessionResponse, SandboxRoute, SandboxesPaginationResponse, Session, SessionAndRoutesResponse, SessionResponse, SessionsResponse, Snapshot, SnapshotResponse, SnapshotTreeNode, SnapshotTreeResponse, SnapshotsResponse, StopSessionResponse, UpdateSandboxResponse, V2NetworkPolicyObjectValidator };
