// @vercel/prepare-flags-definitions/dist/index.js
// src/index.ts
import { createHash } from "crypto";
import { mkdir, writeFile } from "fs/promises";
import { join } from "path";

// package.json
var version = "0.3.0";

// src/index.ts
var FLAGS_HOST = "https://flags.vercel.com";
var FLAGS_DEFINITIONS_VERSION = "1.0.1";
function obfuscate(sdkKey, prefixLength = 18) {
  if (prefixLength >= sdkKey.length) return sdkKey;
  return sdkKey.slice(0, prefixLength) + "*".repeat(sdkKey.length - prefixLength);
}
function hashSdkKey(sdkKey) {
  return createHash("sha256").update(sdkKey).digest("hex");
}
function getProjectIdFromOidcToken(oidcToken) {
  const tokenParts = oidcToken.split(".");
  if (tokenParts.length !== 3 || !tokenParts[1]) {
    return;
  }
  const payload = JSON.parse(
    Buffer.from(tokenParts[1], "base64url").toString("utf8")
  );
  if (typeof payload.project_id !== "string" || !payload.project_id) {
    return;
  }
  return payload.project_id;
}
function generateDefinitionConstants(lines, entries) {
  const stringToConst = /* @__PURE__ */ new Map();
  return entries.map((entry) => {
    const stringified = JSON.stringify(entry.definitions);
    let definitionConst = stringToConst.get(stringified);
    if (!definitionConst) {
      definitionConst = `_d${stringToConst.size}`;
      stringToConst.set(stringified, definitionConst);
      lines.push(
        `const ${definitionConst} = memo(() => JSON.parse(${JSON.stringify(stringified)}));`
      );
    }
    return { key: entry.key, value: definitionConst };
  });
}
function generateMap(lines, entries) {
  lines.push("");
  lines.push("const map = {");
  for (const entry of entries) {
    lines.push(`  ${JSON.stringify(entry.key)}: ${entry.value},`);
  }
  lines.push("};");
  lines.push("");
  lines.push("export function get(key) {");
  lines.push("  return map[key]?.() ?? null;");
  lines.push("}");
}
async function fetchDatafile(token, env, fetchFn, userAgentSuffix) {
  const headers = {
    authorization: `Bearer ${token}`,
    "user-agent": [
      `@vercel/prepare-flags-definitions/${version}`,
      userAgentSuffix
    ].filter(Boolean).join(" ")
  };
  if (env.VERCEL_PROJECT_ID) {
    headers["x-vercel-project-id"] = env.VERCEL_PROJECT_ID;
  }
  if (env.VERCEL_ENV) {
    headers["x-vercel-env"] = env.VERCEL_ENV;
  }
  if (env.VERCEL_DEPLOYMENT_ID) {
    headers["x-vercel-deployment-id"] = env.VERCEL_DEPLOYMENT_ID;
  }
  if (env.VERCEL_REGION) {
    headers["x-vercel-region"] = env.VERCEL_REGION;
  }
  const res = await fetchFn(`${FLAGS_HOST}/v1/datafile`, { headers });
  if (!res.ok) {
    if (res.status === 404) {
      return void 0;
    }
    throw new Error(
      `Failed to fetch flag definitions for ${obfuscate(token)}: ${res.status} ${res.statusText}`
    );
  }
  return res.json();
}
function generateDefinitionsModule(entries, output) {
  output == null ? void 0 : output.debug(
    `vercel-flags: writing flag definitions for "${entries.map(({ key }) => obfuscate(key)).join(", ")}"`
  );
  const lines = [
    "const memo = (fn) => { let cached; return () => (cached ??= fn()); };",
    ""
  ];
  const generatedDefinitions = generateDefinitionConstants(lines, entries);
  generateMap(lines, generatedDefinitions);
  lines.push(
    "",
    `export const version = ${JSON.stringify(FLAGS_DEFINITIONS_VERSION)};`
  );
  return lines.join("\n");
}
var SDK_KEY_REGEX = /^vf_(?:server|client)_/;
function collectFlagEntries(env, output) {
  const entries = [];
  const sdkKeys = Array.from(
    Object.values(env).reduce((acc, value) => {
      if (typeof value === "string") {
        if (SDK_KEY_REGEX.test(value)) {
          acc.add(value);
        } else if (value.startsWith("flags:")) {
          const params = new URLSearchParams(value.slice("flags:".length));
          const sdkKey = params.get("sdkKey");
          if (sdkKey && SDK_KEY_REGEX.test(sdkKey)) {
            acc.add(sdkKey);
          }
        }
      }
      return acc;
    }, /* @__PURE__ */ new Set())
  );
  if (sdkKeys.length > 0) {
    output == null ? void 0 : output.debug(`vercel-flags: found ${sdkKeys.length} SDK keys`);
    for (const key of sdkKeys) {
      entries.push({ type: "sdkKey", key });
    }
  }
  const oidcToken = env.VERCEL_OIDC_TOKEN;
  if (oidcToken && (oidcToken == null ? void 0 : oidcToken.length) > 0) {
    output == null ? void 0 : output.debug(`vercel-flags: found OIDC token`);
    entries.push({ type: "oidcToken", key: oidcToken });
  }
  return entries;
}
async function prepareFlagsDefinitions(options) {
  const {
    cwd,
    env,
    userAgentSuffix,
    fetch: fetchFn = globalThis.fetch,
    output
  } = options;
  output == null ? void 0 : output.debug("vercel-flags: checking env vars for SDK Keys and OIDC Token");
  const entries = collectFlagEntries(env, output);
  if (entries.length === 0) {
    return { created: false, reason: "no-flags-entries" };
  }
  const resolvedEntries = await Promise.all(
    entries.map(async ({ key, type }) => {
      const definitions = await fetchDatafile(
        key,
        env,
        fetchFn,
        userAgentSuffix
      );
      if (!definitions) {
        return;
      }
      if (type === "oidcToken") {
        const projectId = getProjectIdFromOidcToken(key);
        if (projectId) {
          return {
            key: projectId,
            definitions
          };
        }
      }
      if (type === "sdkKey") {
        return {
          key: hashSdkKey(key),
          definitions
        };
      }
    })
  );
  const validEntries = resolvedEntries.filter((entry) => !!entry);
  const definitionsJs = generateDefinitionsModule(validEntries, output);
  const storageDir = join(cwd, "node_modules", "@vercel", "flags-definitions");
  const indexPath = join(storageDir, "index.js");
  const dtsPath = join(storageDir, "index.d.ts");
  const packageJsonPath = join(storageDir, "package.json");
  const dts = [
    "export function get(key: string): Record<string, unknown> | null;",
    "export const version: string;",
    ""
  ].join("\n");
  const packageJson = {
    name: "@vercel/flags-definitions",
    version: FLAGS_DEFINITIONS_VERSION,
    type: "module",
    main: "./index.js",
    types: "./index.d.ts",
    exports: {
      ".": {
        types: "./index.d.ts",
        import: "./index.js"
      }
    }
  };
  await mkdir(storageDir, { recursive: true });
  await Promise.all([
    writeFile(indexPath, definitionsJs),
    writeFile(dtsPath, dts),
    writeFile(packageJsonPath, JSON.stringify(packageJson, null, 2))
  ]);
  output == null ? void 0 : output.debug("vercel-flags: created module");
  output == null ? void 0 : output.debug(`  \u2192 ${indexPath}`);
  output == null ? void 0 : output.debug(`  \u2192 ${dtsPath}`);
  output == null ? void 0 : output.debug(`  \u2192 ${packageJsonPath}`);
  return { created: true, entryCount: entries.length };
}
export {
  generateDefinitionsModule,
  getProjectIdFromOidcToken,
  hashSdkKey,
  prepareFlagsDefinitions
};
