// @vercel/cli-auth/credentials-store.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.resolveCredStorage = resolveCredStorage;
exports.readGlobalConfig = readGlobalConfig;
exports.getCredStorage = getCredStorage;
exports.setLoadKeyringModuleForTesting = setLoadKeyringModuleForTesting;
exports.CredentialsStore = CredentialsStore;
exports.readCredentials = readCredentials;
exports.readCliAuthConfig = readCliAuthConfig;
exports.writeCredentials = writeCredentials;
exports.persistCliAuthConfig = persistCliAuthConfig;
exports.clearCredentials = clearCredentials;
exports.clearAllCredentials = clearAllCredentials;
exports.clearAllCredentialsStrict = clearAllCredentialsStrict;
const node_crypto_1 = require("node:crypto");
const node_assert_1 = __importDefault(require("node:assert"));
const cliConfig = __importStar(require("@vercel/cli-config"));
const KEYRING_SERVICE = 'com.vercel.vercel-cli';
const TOKEN_STORAGE_ENV = 'VERCEL_TOKEN_STORAGE';
const KEYRING_UNAVAILABLE_ENV = 'VERCEL_TEST_KEYRING_UNAVAILABLE';
const KEYRING_UNAVAILABLE_ERROR = 'OS keyring support is unavailable. Set `credStorage` to `auto` or `file` to store credentials in auth.json instead.';
const defaultLoadKeyringModule = () => require('@napi-rs/keyring');
let loadKeyringModule = defaultLoadKeyringModule;
function parseCredentials(value) {
    const { tokenSource, ...credentials } = cliConfig.parseAuthConfig(value);
    return credentials;
}
function formatCredStorageError(value, source) {
    try {
        cliConfig.parseGlobalConfig({ credStorage: value });
    }
    catch (error) {
        if (error instanceof Error) {
            return error.message.replace('`credStorage`', `\`${source}\``);
        }
    }
    return `Invalid value for \`${source}\`: ${JSON.stringify(value)}. Expected one of: ${cliConfig.CRED_STORAGE_CONFIG_VALUES.map(storage => JSON.stringify(storage)).join(', ')}.`;
}
function parseCredStorage(value, source) {
    if (typeof value === 'undefined') {
        return cliConfig.DEFAULT_CRED_STORAGE;
    }
    try {
        const config = cliConfig.parseGlobalConfig({ credStorage: value });
        (0, node_assert_1.default)(config.credStorage !== undefined, 'expected credStorage value to be non-empty');
        return config.credStorage;
    }
    catch {
        throw new Error(formatCredStorageError(value, source));
    }
}
function resolveCredStorage(value) {
    return parseCredStorage(value, 'credStorage');
}
function readRuntimeFlags() {
    return {
        keyringUnavailableForTesting: process.env[KEYRING_UNAVAILABLE_ENV] === '1',
    };
}
function resolveCliAuthConfig(dir, config) {
    const runtimeFlags = readRuntimeFlags();
    if (config?.credStorage) {
        return {
            ...runtimeFlags,
            credStorage: resolveCredStorage(config.credStorage),
        };
    }
    const credStorageOverride = process.env[TOKEN_STORAGE_ENV];
    if (typeof credStorageOverride !== 'undefined') {
        return {
            ...runtimeFlags,
            credStorage: parseCredStorage(credStorageOverride, TOKEN_STORAGE_ENV),
        };
    }
    return {
        ...runtimeFlags,
        credStorage: resolveCredStorage(readGlobalConfig(dir).credStorage),
    };
}
function readGlobalConfig(dir) {
    const configPath = cliConfig.getConfigFilePath(dir);
    try {
        const config = cliConfig.readGlobalConfigFile(configPath);
        return {
            credStorage: typeof config.credStorage === 'undefined'
                ? undefined
                : resolveCredStorage(config.credStorage),
        };
    }
    catch (error) {
        if (isENOENTError(error)) {
            return {};
        }
        throw error;
    }
}
function getCredStorage(dir, config) {
    return resolveCliAuthConfig(dir, config).credStorage;
}
/** Return an account key suitable for use in keyrings for a given configPath.
 *
 *  `auth.json` and the keychain are supposed to be equivalent and exchangeable.
 * Since file-based credentials storage is configurable (via config path),
 * we must ensure that keyring entries are keyed by config path also.  We cannot
 * use the full path as the key, so use a 16-byte prefix of sha256 hash of
 * the path.
 */
function getKeyringAccount(configPath) {
    const digest = (0, node_crypto_1.createHash)('sha256').update(configPath).digest('hex');
    return `cli:${digest.slice(0, 16)}`;
}
function setLoadKeyringModuleForTesting(nextLoadKeyringModule) {
    loadKeyringModule = nextLoadKeyringModule ?? defaultLoadKeyringModule;
}
function createKeyringEntry(configPath, runtimeFlags) {
    try {
        if (runtimeFlags.keyringUnavailableForTesting) {
            throw new Error(`${KEYRING_UNAVAILABLE_ENV} is set`);
        }
        const { Entry } = loadKeyringModule();
        return new Entry(KEYRING_SERVICE, getKeyringAccount(configPath));
    }
    catch (error) {
        const wrappedError = new Error(KEYRING_UNAVAILABLE_ERROR);
        wrappedError.cause = error;
        throw wrappedError;
    }
}
function isKeyringUnavailableError(error) {
    return error instanceof Error && error.message === KEYRING_UNAVAILABLE_ERROR;
}
function readCredentialsFromFile(configPath) {
    return cliConfig.readAuthFileConfig(configPath);
}
function isENOENTError(error) {
    return (typeof error === 'object' &&
        error !== null &&
        'code' in error &&
        error.code === 'ENOENT');
}
function createMissingCredentialsError(configPath) {
    const error = new Error(`ENOENT: no such file or directory, open '${configPath}'`);
    Object.assign(error, { code: 'ENOENT' });
    return error;
}
function writeCredentialsToFile(configPath, config) {
    cliConfig.writeAuthConfigFile(configPath, config);
}
function readCredentialsFromKeyring(configPath, runtimeFlags) {
    const entry = createKeyringEntry(configPath, runtimeFlags);
    const value = entry.getPassword();
    if (value === null) {
        return null;
    }
    return parseCredentials(JSON.parse(value));
}
function writeCredentialsToKeyring(configPath, config, runtimeFlags) {
    const entry = createKeyringEntry(configPath, runtimeFlags);
    entry.setPassword(JSON.stringify(config));
}
function deleteCredentialsFromKeyring(configPath, runtimeFlags) {
    const entry = createKeyringEntry(configPath, runtimeFlags);
    return entry.deleteCredential();
}
function deleteCredentialsFile(configPath) {
    cliConfig.deleteAuthConfigFile(configPath);
}
function tryDeleteCredentialsFromKeyring(configPath, runtimeFlags) {
    try {
        deleteCredentialsFromKeyring(configPath, runtimeFlags);
    }
    catch { }
}
function deleteCredentialsFromKeyringIfAvailable(configPath, runtimeFlags) {
    try {
        deleteCredentialsFromKeyring(configPath, runtimeFlags);
    }
    catch (error) {
        if (!isKeyringUnavailableError(error)) {
            throw error;
        }
    }
}
function migrateCredentialsToFile(configPath, credentials, runtimeFlags) {
    writeCredentialsToFile(configPath, credentials);
    tryDeleteCredentialsFromKeyring(configPath, runtimeFlags);
    return credentials;
}
function tryReadKeyringCredentialsIntoFile(configPath, runtimeFlags) {
    try {
        const credentials = readCredentialsFromKeyring(configPath, runtimeFlags);
        if (credentials === null) {
            return null;
        }
        return migrateCredentialsToFile(configPath, credentials, runtimeFlags);
    }
    catch {
        return null;
    }
}
function readFileStorageCredentials(configPath, runtimeFlags) {
    try {
        const credentials = readCredentialsFromFile(configPath);
        if (cliConfig.authConfigHasUsableTokenData(credentials)) {
            return credentials;
        }
        return (tryReadKeyringCredentialsIntoFile(configPath, runtimeFlags) ?? credentials);
    }
    catch (error) {
        if (!isENOENTError(error)) {
            throw error;
        }
        const credentials = tryReadKeyringCredentialsIntoFile(configPath, runtimeFlags);
        if (credentials !== null) {
            return credentials;
        }
        throw error;
    }
}
function migrateCredentialsToKeyring(configPath, credentials, runtimeFlags) {
    writeCredentialsToKeyring(configPath, credentials, runtimeFlags);
    deleteCredentialsFile(configPath);
    return credentials;
}
function readKeyringStorageCredentials(configPath, runtimeFlags) {
    const credentials = readCredentialsFromKeyring(configPath, runtimeFlags);
    if (credentials !== null) {
        return credentials;
    }
    try {
        const fileCredentials = readCredentialsFromFile(configPath);
        return migrateCredentialsToKeyring(configPath, fileCredentials, runtimeFlags);
    }
    catch (error) {
        if (isENOENTError(error)) {
            throw createMissingCredentialsError(configPath);
        }
        throw error;
    }
}
function readAutoStorageCredentials(configPath, runtimeFlags) {
    try {
        const credentials = readCredentialsFromKeyring(configPath, runtimeFlags);
        if (credentials !== null) {
            return credentials;
        }
    }
    catch { }
    return readCredentialsFromFile(configPath);
}
function writeFileStorageCredentials(configPath, credentials, runtimeFlags) {
    writeCredentialsToFile(configPath, credentials);
    tryDeleteCredentialsFromKeyring(configPath, runtimeFlags);
    return 'file';
}
function writeKeyringStorageCredentials(configPath, credentials, runtimeFlags) {
    writeCredentialsToKeyring(configPath, credentials, runtimeFlags);
    deleteCredentialsFile(configPath);
    return 'keyring';
}
function writeAutoStorageCredentials(configPath, credentials, runtimeFlags) {
    try {
        return writeKeyringStorageCredentials(configPath, credentials, runtimeFlags);
    }
    catch {
        deleteCredentialsFromKeyringIfAvailable(configPath, runtimeFlags);
        return writeFileStorageCredentials(configPath, credentials, runtimeFlags);
    }
}
function deleteFileStorageCredentials(configPath, runtimeFlags) {
    deleteCredentialsFile(configPath);
    deleteCredentialsFromKeyringIfAvailable(configPath, runtimeFlags);
}
function deleteKeyringStorageCredentials(configPath, runtimeFlags) {
    deleteCredentialsFromKeyring(configPath, runtimeFlags);
    deleteCredentialsFile(configPath);
}
function deleteAutoStorageCredentials(configPath, runtimeFlags) {
    try {
        deleteCredentialsFromKeyring(configPath, runtimeFlags);
    }
    catch { }
    deleteCredentialsFile(configPath);
}
function CredentialsStore(dir, config = {}) {
    const configPath = cliConfig.getAuthConfigFilePath(dir);
    const cliAuthConfig = resolveCliAuthConfig(dir, config);
    const tokenStorage = cliAuthConfig.credStorage;
    return {
        configPath,
        get() {
            switch (tokenStorage) {
                case 'file':
                    return readFileStorageCredentials(configPath, cliAuthConfig);
                case 'keyring':
                    return readKeyringStorageCredentials(configPath, cliAuthConfig);
                case 'auto':
                    return readAutoStorageCredentials(configPath, cliAuthConfig);
            }
            throw new Error(`Unsupported credStorage: ${tokenStorage}`);
        },
        update(config) {
            if (config.skipWrite)
                return;
            const parsedConfig = parseCredentials(config);
            switch (tokenStorage) {
                case 'file':
                    return writeFileStorageCredentials(configPath, parsedConfig, cliAuthConfig);
                case 'keyring':
                    return writeKeyringStorageCredentials(configPath, parsedConfig, cliAuthConfig);
                case 'auto':
                    return writeAutoStorageCredentials(configPath, parsedConfig, cliAuthConfig);
            }
            throw new Error(`Unsupported credStorage: ${tokenStorage}`);
        },
        delete() {
            switch (tokenStorage) {
                case 'file':
                    deleteFileStorageCredentials(configPath, cliAuthConfig);
                    break;
                case 'keyring':
                    deleteKeyringStorageCredentials(configPath, cliAuthConfig);
                    break;
                case 'auto':
                    deleteAutoStorageCredentials(configPath, cliAuthConfig);
                    break;
            }
        },
    };
}
function readCredentials(dir, config = {}) {
    return CredentialsStore(dir, config).get();
}
function readCliAuthConfig(dir, config = {}) {
    return readCredentials(dir, config);
}
function writeCredentials(dir, credentials, config = {}) {
    return CredentialsStore(dir, config).update(credentials);
}
function toPersistedAuthConfig(authConfig) {
    const { skipWrite, tokenSource, ...persistedAuthConfig } = authConfig;
    return persistedAuthConfig;
}
function persistCliAuthConfig(dir, authConfig, config = {}) {
    if (authConfig.skipWrite) {
        return;
    }
    const persistedAuthConfig = toPersistedAuthConfig(authConfig);
    if (!cliConfig.authConfigHasUsableTokenData(persistedAuthConfig)) {
        clearAllCredentialsStrict(dir);
        return;
    }
    return writeCredentials(dir, persistedAuthConfig, config);
}
function clearCredentials(dir, config = {}) {
    CredentialsStore(dir, config).delete();
}
function clearAllCredentials(dir) {
    for (const credStorage of cliConfig.CRED_STORAGE_VALUES) {
        try {
            clearCredentials(dir, { credStorage });
        }
        catch { }
    }
}
function clearAllCredentialsStrict(dir) {
    const configPath = cliConfig.getAuthConfigFilePath(dir);
    const runtimeFlags = readRuntimeFlags();
    let fileError;
    let keyringError;
    try {
        deleteCredentialsFile(configPath);
    }
    catch (error) {
        fileError = error;
    }
    try {
        deleteCredentialsFromKeyring(configPath, runtimeFlags);
    }
    catch (error) {
        if (!isKeyringUnavailableError(error)) {
            keyringError = error;
        }
    }
    if (keyringError) {
        throw keyringError;
    }
    if (fileError) {
        throw fileError;
    }
}
