// @vercel/cli-auth/sso.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.reauthorizeTeam = reauthorizeTeam;
exports.parseSSOError = parseSSOError;
const node_http_1 = __importDefault(require("node:http"));
const async_listen_1 = require("async-listen");
const open_1 = __importDefault(require("open"));
const z = __importStar(require("zod/mini"));
/**
 * Re-authorize a team's SSO session by opening the browser and waiting for the user to complete
 * the SSO flow. Use when the user's SSO session has expired or missing.
 *
 */
async function reauthorizeTeam(params) {
    const { team, token, oauth } = params;
    if (!token) {
        throw new TypeError('Token is missing, cannot initiate SSO re-authorization. Please log in again.');
    }
    const introspection = await oauth.introspectToken(token);
    const { active, session_id, client_id } = introspection;
    if (!active || !session_id || !client_id) {
        throw new Error('Introspection failed. Token is either expired or otherwise invalid. Please log in again.');
    }
    const verificationToken = await waitForVerification({
        team,
        session_id,
        client_id,
    });
    await fetch(`https://api.vercel.com/registration/verify?${new URLSearchParams({ token: verificationToken })}`);
}
/**
 * Get a verification token by spawning a localhost HTTP server
 * that gets redirected to as the OAuth callback URL.
 *
 * The token returned can be sent back to the Vercel API to complete
 * the SSO re-authorization.
 */
async function waitForVerification(params) {
    const { team, session_id, client_id } = params;
    const server = node_http_1.default.createServer();
    const { port } = await (0, async_listen_1.listen)(server, 0, '127.0.0.1');
    const initSSOUrl = new URL(`/sso/${team}?${new URLSearchParams({
        session_id,
        client_id,
        next: `http://localhost:${port}`,
    })}`, `https://vercel.com`);
    try {
        const [query] = await Promise.all([
            new Promise((resolve, reject) => {
                server.once('request', (req, res) => {
                    // Close the HTTP connection to prevent `server.close()` from hanging
                    res.setHeader('connection', 'close');
                    const { searchParams } = new URL(req.url || '/', 'http://localhost');
                    resolve(searchParams);
                    // Redirect the user back to the Vercel login notification page
                    res.statusCode = 302;
                    res.setHeader('location', getNotificationUrl(searchParams)).end();
                });
                server.once('error', reject);
            }),
            (0, open_1.default)(initSSOUrl.href),
        ]);
        const loginError = query.get('loginError');
        if (loginError) {
            const error = new Error('Login failed.');
            error.cause = JSON.parse(loginError);
            throw error;
        }
        const token = query.get('token');
        if (!token) {
            throw new TypeError('Could not get verification token. Please contact support.');
        }
        return token;
    }
    finally {
        server.close();
    }
}
function getNotificationUrl(params) {
    const loginState = params.has('loginError')
        ? 'failed'
        : params.has('ssoEmail')
            ? 'incomplete'
            : 'success';
    return `https://vercel.com/notifications/cli-login-${loginState}?${params}`;
}
const SSOError = z.object({
    code: z.string(),
    message: z.string(),
    scope: z.optional(z.string()),
    enforced: z.optional(z.boolean()),
});
/** Parse an error-like object to determine if SSO re-authorization is required. */
function parseSSOError(error) {
    const { data } = SSOError.safeParse(error);
    if (!data?.enforced)
        return;
    return data;
}
