// @vercel/sandbox/dist/snapshot.cjs
const require_rolldown_runtime = require('./_virtual/rolldown_runtime.cjs');
const require_api_client = require('./api-client/api-client.cjs');
require('./api-client/index.cjs');
const require_get_credentials = require('./utils/get-credentials.cjs');
const require_paginator = require('./utils/paginator.cjs');
let __workflow_serde = require("@workflow/serde");

//#region src/snapshot.ts
/**
* A Snapshot is a saved state of a Sandbox that can be used to create new Sandboxes
*
* Use {@link Sandbox.snapshot} or {@link Snapshot.get} to construct.
* @hideconstructor
*/
var Snapshot = class Snapshot {
	/**
	* Lazily resolve credentials and construct an API client.
	* This is used in step contexts where the Snapshot was deserialized
	* without a client (e.g. when crossing workflow/step boundaries).
	* @internal
	*/
	async ensureClient() {
		"use step";
		if (this._client) return this._client;
		const credentials = await require_get_credentials.getCredentials();
		this._client = new require_api_client.APIClient({
			teamId: credentials.teamId,
			token: credentials.token
		});
		return this._client;
	}
	/**
	* Unique ID of this snapshot.
	*/
	get snapshotId() {
		return this.snapshot.id;
	}
	/**
	* The ID of the session from which this snapshot was created.
	*/
	get sourceSessionId() {
		return this.snapshot.sourceSessionId;
	}
	/**
	* The status of the snapshot.
	*/
	get status() {
		return this.snapshot.status;
	}
	/**
	* The size of the snapshot in bytes, or null if not available.
	*/
	get sizeBytes() {
		return this.snapshot.sizeBytes;
	}
	/**
	* The creation date of this snapshot.
	*/
	get createdAt() {
		return new Date(this.snapshot.createdAt);
	}
	/**
	* When this snapshot was last updated.
	*/
	get updatedAt() {
		return new Date(this.snapshot.updatedAt);
	}
	/**
	* The expiration date of this snapshot, or undefined if it does not expire.
	*/
	get expiresAt() {
		if (this.snapshot.expiresAt === void 0) return;
		return new Date(this.snapshot.expiresAt);
	}
	/**
	* Serialize a Snapshot instance to plain data for @workflow/serde.
	*
	* @param instance - The Snapshot instance to serialize
	* @returns A plain object containing snapshot metadata
	*/
	static [__workflow_serde.WORKFLOW_SERIALIZE](instance) {
		return { snapshot: instance.snapshot };
	}
	/**
	* Deserialize a Snapshot from serialized data.
	*
	* The deserialized instance uses the serialized metadata synchronously and
	* lazily creates an API client only when methods perform API requests.
	*
	* @param data - The serialized snapshot data
	* @returns The reconstructed Snapshot instance
	*/
	static [__workflow_serde.WORKFLOW_DESERIALIZE](data) {
		return new Snapshot({ snapshot: data.snapshot });
	}
	constructor({ client, snapshot }) {
		this._client = null;
		this._client = client ?? null;
		this.snapshot = snapshot;
	}
	/**
	* Allow to get a list of snapshots for a team narrowed to the given params.
	* It returns both the snapshots and the pagination metadata to allow getting
	* the next page of results.
	*
	* The returned object is async-iterable to auto-paginate through all pages:
	*
	* ```ts
	* const result = await Snapshot.list({ name: "my-sandbox" });
	* for await (const snapshot of result) { ... }
	* // or: await result.toArray();
	* // or: for await (const page of result.pages()) { ... }
	* ```
	*/
	static async list(params) {
		"use step";
		const credentials = await require_get_credentials.getCredentials(params);
		const client = new require_api_client.APIClient({
			teamId: credentials.teamId,
			token: credentials.token,
			fetch: params?.fetch
		});
		const fetchPage = async (cursor) => {
			return (await client.listSnapshots({
				...credentials,
				...params,
				...cursor !== void 0 && { cursor }
			})).json;
		};
		return require_paginator.attachPaginator(await fetchPage(params?.cursor), {
			itemsKey: "snapshots",
			fetchNext: fetchPage,
			signal: params?.signal
		});
	}
	/**
	* Fetch the snapshot ancestry tree anchored on a given snapshot.
	* It returns both the tree nodes and the pagination metadata to allow
	* walking the next page of results in the same direction.
	*
	* The returned object is async-iterable to auto-paginate through all pages
	* in the direction set by `sortOrder` (`"desc"` walks ancestors, `"asc"`
	* walks descendants):
	*
	* ```ts
	* const result = await Snapshot.tree({ snapshotId: "snap_abc", sortOrder: "desc" });
	* for await (const node of result) { ... }
	* // or: await result.toArray();
	* // or: for await (const page of result.pages()) { ... }
	* ```
	*/
	static async tree(params) {
		"use step";
		const credentials = await require_get_credentials.getCredentials(params);
		const client = new require_api_client.APIClient({
			teamId: credentials.teamId,
			token: credentials.token,
			fetch: params.fetch
		});
		const fetchPage = async (snapshotId) => {
			return (await client.getSnapshotTree({
				...credentials,
				...params,
				snapshotId
			})).json;
		};
		return require_paginator.attachPaginator(await fetchPage(params.snapshotId), {
			itemsKey: "snapshots",
			fetchNext: fetchPage,
			signal: params.signal
		});
	}
	/**
	* Retrieve an existing snapshot.
	*
	* @param params - Get parameters and optional credentials.
	* @returns A promise resolving to the {@link Sandbox}.
	*/
	static async get(params) {
		"use step";
		const credentials = await require_get_credentials.getCredentials(params);
		const client = new require_api_client.APIClient({
			teamId: credentials.teamId,
			token: credentials.token
		});
		return new Snapshot({
			client,
			snapshot: (await client.getSnapshot({
				snapshotId: params.snapshotId,
				signal: params.signal
			})).json.snapshot
		});
	}
	/**
	* Delete this snapshot.
	*
	* @param opts - Optional parameters.
	* @param opts.signal - An AbortSignal to cancel the operation.
	* @returns A promise that resolves once the snapshot has been deleted.
	*/
	async delete(opts) {
		"use step";
		this.snapshot = (await (await this.ensureClient()).deleteSnapshot({
			snapshotId: this.snapshot.id,
			signal: opts?.signal
		})).json.snapshot;
	}
};

//#endregion
exports.Snapshot = Snapshot;
