// @vercel/backends/dist/rolldown/index.mjs
import { builtinModules } from "node:module";
import { FileBlob, FileFsRef, Span, isBackendFramework } from "@vercel/build-utils";
import { dirname, extname, isAbsolute, join, posix, relative, resolve, sep } from "node:path";
import { existsSync } from "node:fs";
import { lstat, readFile, readlink, stat } from "node:fs/promises";
import { build } from "rolldown";
import { nodeFileTrace } from "@vercel/nft";
import { isNativeError } from "node:util/types";
import { transform } from "oxc-transform";
import { exports } from "resolve.exports";

//#region src/cervel/utils.ts
const noColor = globalThis.process?.env?.NO_COLOR === "1" || globalThis.process?.env?.TERM === "dumb";
const resets = {
	1: 22,
	31: 39,
	32: 39,
	33: 39,
	34: 39,
	35: 39,
	36: 39,
	90: 39
};
const _c = (c) => (text) => {
	if (noColor) return text;
	return `\u001B[${c}m${text}\u001B[${resets[c] ?? 0}m`;
};
const Colors = {
	bold: _c(1),
	red: _c(31),
	green: _c(32),
	yellow: _c(33),
	blue: _c(34),
	magenta: _c(35),
	cyan: _c(36),
	gray: _c(90),
	url: (title, url) => noColor ? `[${title}](${url})` : `\u001B]8;;${url}\u001B\\${title}\u001B]8;;\u001B\\`
};

//#endregion
//#region src/rolldown/resolve-format.ts
const resolveEntrypointAndFormat = async (args) => {
	const extension = extname(args.entrypoint);
	const extensionMap = {
		".ts": {
			format: "auto",
			extension: "js"
		},
		".mts": {
			format: "esm",
			extension: "mjs"
		},
		".cts": {
			format: "cjs",
			extension: "cjs"
		},
		".cjs": {
			format: "cjs",
			extension: "cjs"
		},
		".js": {
			format: "auto",
			extension: "js"
		},
		".mjs": {
			format: "esm",
			extension: "mjs"
		}
	};
	const extensionInfo = extensionMap[extension] || extensionMap[".js"];
	let resolvedFormat = extensionInfo.format === "auto" ? args.defaultFormat : extensionInfo.format;
	const packageJsonPath = join(args.workPath, "package.json");
	let pkg = {};
	if (existsSync(packageJsonPath)) {
		const source = await readFile(packageJsonPath, "utf8");
		try {
			pkg = JSON.parse(source.toString());
		} catch (_e) {
			pkg = {};
		}
		if (extensionInfo.format === "auto") if (pkg.type === "module") resolvedFormat = "esm";
		else resolvedFormat = "cjs";
	}
	if (!resolvedFormat) resolvedFormat = "esm";
	return {
		format: resolvedFormat,
		extension: resolvedFormat === "esm" ? "mjs" : "cjs"
	};
};

//#endregion
//#region src/rolldown/nft.ts
const nft = async (args) => {
	const nftSpan = args.span.child("vc.builder.backends.nft");
	const runNft = async () => {
		const virtualFiles = /* @__PURE__ */ new Map();
		if (args.traceFiles) for (const [relPath, file] of Object.entries(args.files)) {
			if (!isJsLikeExtension(relPath) || file.type !== "FileBlob") continue;
			virtualFiles.set(join(args.repoRootPath, relPath), typeof file.data === "string" ? file.data : Buffer.from(new Uint8Array(file.data)));
		}
		const traceRoots = [...Array.from(args.localBuildFiles).filter((p) => existsSync(p) || virtualFiles.has(p)), ...virtualFiles.keys()];
		const traceBase = getCommonBase(args.repoRootPath, traceRoots);
		const ignorePatterns = getIgnorePatterns({
			ignoreNodeModules: args.ignoreNodeModules,
			ignore: args.ignore,
			traceBase,
			repoRootPath: args.repoRootPath
		});
		const statOverride = memoize(async (fsPath) => {
			const virtual = virtualFiles.get(fsPath);
			if (virtual !== void 0) return createVirtualFileStat(virtual);
			try {
				return await stat(fsPath);
			} catch (error) {
				if (isNativeError(error) && "code" in error && (error.code === "ENOENT" || error.code === "ENOTDIR")) return null;
				throw error;
			}
		});
		const readlinkOverride = memoize(async (fsPath) => {
			if (virtualFiles.has(fsPath)) return null;
			try {
				return await readlink(fsPath);
			} catch (error) {
				if (isNativeError(error) && "code" in error && (error.code === "EINVAL" || error.code === "ENOENT" || error.code === "ENOTDIR")) return null;
				throw error;
			}
		});
		const readFileOverride = memoize(async (fsPath) => {
			const virtual = virtualFiles.get(fsPath);
			if (virtual !== void 0) return virtual;
			try {
				let source = await readFile(fsPath);
				if (isTypeScriptFile(fsPath)) source = (await transform(fsPath, source.toString())).code;
				return source;
			} catch (error) {
				if (isNativeError(error) && "code" in error && (error.code === "ENOENT" || error.code === "EISDIR")) return null;
				throw error;
			}
		});
		const nftResult = await nodeFileTrace(traceRoots, {
			base: traceBase,
			processCwd: args.workPath,
			ts: true,
			mixedModules: true,
			moduleSyncCatchall: true,
			conditions: args.conditions,
			ignore: ignorePatterns.length > 0 ? ignorePatterns : void 0,
			readFile: readFileOverride,
			...virtualFiles.size > 0 ? {
				stat: statOverride,
				readlink: readlinkOverride
			} : {}
		});
		for (const file of nftResult.fileList) {
			const absolutePath = join(traceBase, file);
			if (virtualFiles.has(absolutePath)) continue;
			let stats;
			try {
				stats = await lstat(absolutePath);
			} catch (error) {
				if (isNativeError(error) && "code" in error && error.code === "ENOENT") continue;
				throw error;
			}
			const normalizedFile = normalizePath(relative(args.repoRootPath, absolutePath));
			const outputPath = stripParentSegments(normalizedFile);
			const escapesBase = outputPath !== normalizedFile;
			if (args.localBuildFiles.has(join(args.repoRootPath, outputPath)) && !isNodeModulesPath(outputPath)) continue;
			if (stats.isSymbolicLink() || stats.isFile()) {
				if (stats.isSymbolicLink()) {
					const symlinkTarget = await readlink(absolutePath);
					const symlinkTargetPath = normalizePath(relative(args.repoRootPath, resolve(dirname(absolutePath), symlinkTarget)));
					if (isParentPath(symlinkTargetPath)) {
						const outputTargetPath = stripParentSegments(symlinkTargetPath);
						args.files[outputPath] = new FileBlob({
							data: posix.relative(posix.dirname(outputPath), outputTargetPath),
							mode: stats.mode
						});
						continue;
					}
				}
				if (args.ignoreNodeModules || escapesBase) {
					if ((stats.isSymbolicLink() ? await stat(absolutePath) : stats).isFile()) args.files[outputPath] = new FileBlob({
						data: await readFile(absolutePath),
						mode: stats.mode
					});
				} else args.files[outputPath] = new FileFsRef({
					fsPath: absolutePath,
					mode: stats.mode
				});
			}
		}
	};
	await nftSpan.trace(runNft);
};
/** Memoizes an async fs lookup by path, caching negative/ENOENT results too. */
const memoize = (fn) => {
	const cache = /* @__PURE__ */ new Map();
	return (path) => {
		let cached = cache.get(path);
		if (cached === void 0) {
			cached = fn(path);
			cache.set(path, cached);
		}
		return cached;
	};
};
const JS_LIKE_EXTENSIONS = new Set([
	".js",
	".cjs",
	".mjs",
	".ts",
	".cts",
	".mts",
	".tsx",
	".jsx",
	".json",
	".node"
]);
const isJsLikeExtension = (path) => {
	const dot = path.lastIndexOf(".");
	if (dot === -1) return false;
	return JS_LIKE_EXTENSIONS.has(path.slice(dot).toLowerCase());
};
const getCommonBase = (base, paths) => {
	let commonBase = base;
	for (const path of paths) while (!isPathInside(commonBase, path)) {
		const parent = dirname(commonBase);
		if (parent === commonBase) break;
		commonBase = parent;
	}
	return commonBase;
};
const isPathInside = (base, path) => {
	const relPath = relative(base, path);
	return relPath === "" || !isParentPath(normalizePath(relPath)) && !isAbsolute(relPath);
};
const normalizePath = (path) => path.split(sep).join("/");
const isParentPath = (path) => path === ".." || path.startsWith("../");
const isNodeModulesPath = (path) => path.split("/").includes("node_modules");
const stripParentSegments = (path) => {
	const segments = path.split("/");
	let index = 0;
	while (segments[index] === "..") index++;
	return segments.slice(index).join("/");
};
const getIgnorePatterns = ({ ignoreNodeModules, ignore, traceBase, repoRootPath }) => {
	const patterns = [...ignoreNodeModules ? ["**/node_modules/**"] : [], ...ignore ? Array.isArray(ignore) ? ignore : [ignore] : []];
	const repoRootRelativeToTraceBase = normalizePath(relative(traceBase, repoRootPath));
	if (!repoRootRelativeToTraceBase || isParentPath(repoRootRelativeToTraceBase)) return patterns;
	const rebasedPatterns = patterns.map((pattern) => {
		const isNegated = pattern.startsWith("!");
		const patternBody = isNegated ? pattern.slice(1) : pattern;
		const rebased = posix.join(repoRootRelativeToTraceBase, patternBody);
		return isNegated ? `!${rebased}` : rebased;
	});
	return [...patterns, ...rebasedPatterns];
};
const createVirtualFileStat = (data) => {
	const now = /* @__PURE__ */ new Date();
	return {
		isFile: () => true,
		isDirectory: () => false,
		isBlockDevice: () => false,
		isCharacterDevice: () => false,
		isSymbolicLink: () => false,
		isFIFO: () => false,
		isSocket: () => false,
		dev: 0,
		ino: 0,
		mode: 33188,
		nlink: 1,
		uid: 0,
		gid: 0,
		rdev: 0,
		size: typeof data === "string" ? Buffer.byteLength(data) : data.length,
		blksize: 4096,
		blocks: 1,
		atimeMs: now.getTime(),
		mtimeMs: now.getTime(),
		ctimeMs: now.getTime(),
		birthtimeMs: now.getTime(),
		atime: now,
		mtime: now,
		ctime: now,
		birthtime: now
	};
};
const isTypeScriptFile = (fsPath) => {
	if (fsPath.endsWith(".d.ts") || fsPath.endsWith(".d.mts") || fsPath.endsWith(".d.cts")) return false;
	return fsPath.endsWith(".ts") || fsPath.endsWith(".tsx") || fsPath.endsWith(".mts") || fsPath.endsWith(".cts");
};

//#endregion
//#region src/rolldown/index.ts
const PLUGIN_NAME = "vercel:backends";
const CJS_SHIM_PREFIX = "\0cjs-shim:";
const rolldown = async (args) => {
	const files = {};
	const { format, extension } = await resolveEntrypointAndFormat(args);
	const localBuildFiles = /* @__PURE__ */ new Set();
	let handler = null;
	const packageJsonCache = /* @__PURE__ */ new Map();
	const shimMeta = /* @__PURE__ */ new Map();
	const framework = {
		slug: "",
		version: ""
	};
	const getPackageJson = async (pkgPath) => {
		if (packageJsonCache.has(pkgPath)) return packageJsonCache.get(pkgPath);
		try {
			const contents = await readFile(pkgPath, "utf-8");
			const parsed = JSON.parse(contents);
			packageJsonCache.set(pkgPath, parsed);
			return parsed;
		} catch {
			packageJsonCache.set(pkgPath, null);
			return null;
		}
	};
	const isCommonJS = async (bareImport, resolvedPath, resolvedInfo) => {
		const ext = extname(resolvedPath);
		if (ext === ".cjs") return true;
		if (ext === ".mjs") return false;
		if (ext === ".js" || ext === ".ts") {
			const pkgJsonPath = resolvedInfo.packageJsonPath;
			if (!pkgJsonPath) return true;
			const pkgJson = await getPackageJson(pkgJsonPath);
			if (!pkgJson) return true;
			const pkgDir = dirname(pkgJsonPath);
			const relativePath = resolvedPath.slice(pkgDir.length + 1).replace(/\\/g, "/");
			const pkgName = pkgJson.name || "";
			const subpath = bareImport.startsWith(pkgName) ? `.${bareImport.slice(pkgName.length)}` || "." : ".";
			try {
				if (exports(pkgJson, subpath, {
					require: false,
					conditions: ["node", "import"]
				})?.some((p) => p === relativePath || p === `./${relativePath}`)) return false;
				if (exports(pkgJson, subpath, {
					require: true,
					conditions: ["node", "require"]
				})?.some((p) => p === relativePath || p === `./${relativePath}`)) return true;
			} catch {}
			if (pkgJson.module) return false;
			return pkgJson.type !== "module";
		}
		return true;
	};
	const isBareImport = (id) => {
		return !id.startsWith(".") && !id.startsWith("/") && !/^[a-z][a-z0-9+.-]*:/i.test(id);
	};
	const isNodeModule = (resolved) => {
		return resolved?.id?.includes("node_modules") ?? false;
	};
	const isNodeBuiltin = (id) => {
		const normalizedId = id.includes(":") ? id.split(":")[1] : id;
		return builtinModules.includes(normalizedId);
	};
	const isBunBuiltin = (id) => {
		return id === "bun" || id.startsWith("bun:");
	};
	const isLocalImport = (id) => {
		return !id.startsWith("node:") && !id.includes("node_modules");
	};
	const plugin = () => {
		return {
			name: PLUGIN_NAME,
			resolveId: {
				order: "pre",
				async handler(id, importer, rOpts) {
					if (id.startsWith(CJS_SHIM_PREFIX)) return {
						id,
						external: false
					};
					const resolved = await this.resolve(id, importer, rOpts);
					if (isNodeBuiltin(id)) return {
						id: id.startsWith("node:") ? id : `node:${id}`,
						external: true
					};
					if (isBunBuiltin(id)) return {
						id,
						external: true
					};
					if (resolved?.id && isLocalImport(resolved.id)) {
						if (existsSync(resolved.id)) localBuildFiles.add(resolved.id);
					} else if (!resolved && !(isBareImport(id) && importer)) {
						const candidate = join(args.workPath, id);
						if (existsSync(candidate)) localBuildFiles.add(candidate);
					}
					if (importer?.startsWith(CJS_SHIM_PREFIX) && isBareImport(id)) return {
						id,
						external: true
					};
					if (importer && isBareImport(id) && isNodeModule(resolved)) {
						if (isBackendFramework(id) && resolved?.packageJsonPath) try {
							const pkg = await readFile(resolved.packageJsonPath, "utf8");
							const pkgJson = JSON.parse(pkg);
							framework.slug = pkgJson.name;
							framework.version = pkgJson.version;
						} catch {}
						if ((resolved ? await isCommonJS(id, resolved.id, resolved) : false) && resolved) {
							localBuildFiles.add(resolved.id);
							const importerPkgJsonPath = (await this.resolve(importer))?.packageJsonPath;
							if (importerPkgJsonPath) {
								const importerPkgDir = relative(args.repoRootPath, dirname(importerPkgJsonPath));
								const shimId$1 = `${CJS_SHIM_PREFIX}${importerPkgDir.replace(/\//g, "_")}_${id.replace(/\//g, "_")}`;
								shimMeta.set(shimId$1, {
									pkgDir: importerPkgDir,
									pkgName: id
								});
								return {
									id: shimId$1,
									external: false
								};
							}
							const shimId = `${CJS_SHIM_PREFIX}${id.replace(/\//g, "_")}`;
							shimMeta.set(shimId, {
								pkgDir: "",
								pkgName: id
							});
							return {
								id: shimId,
								external: false
							};
						}
						return {
							id,
							external: true
						};
					}
					if (importer && isBareImport(id)) return resolved;
					if (resolved && !isNodeModule(resolved)) return resolved;
					return resolved;
				}
			},
			load: { async handler(id) {
				if (id.startsWith(CJS_SHIM_PREFIX)) {
					const meta = shimMeta.get(id);
					if (!meta) return { code: `module.exports = require('${id.slice(10)}');` };
					const { pkgDir, pkgName } = meta;
					return { code: `
import { createRequire } from 'node:module';
import { fileURLToPath } from 'node:url';
import { dirname, join } from 'node:path';

const requireFromContext = createRequire(join(dirname(fileURLToPath(import.meta.url)), '${pkgDir ? join("..", pkgDir, "package.json") : "../package.json"}'));
module.exports = requireFromContext('${pkgName}');
`.trim() };
				}
				return null;
			} }
		};
	};
	const runRolldown = () => build({
		input: args.entrypoint,
		write: false,
		cwd: args.workPath,
		platform: "node",
		transform: { define: format === "esm" ? {
			__dirname: "import.meta.dirname",
			__filename: "import.meta.filename"
		} : void 0 },
		tsconfig: true,
		plugins: [plugin()],
		output: {
			cleanDir: true,
			format,
			entryFileNames: `[name].${extension}`,
			preserveModules: true,
			preserveModulesRoot: args.repoRootPath,
			sourcemap: false
		}
	});
	const rolldownSpan = args.span?.child("vc.builder.backends.rolldown");
	const out = await rolldownSpan?.trace(runRolldown) || await runRolldown();
	for (const file of out.output) if (file.type === "chunk") {
		if (file.isEntry) handler = file.fileName;
		files[file.fileName] = new FileBlob({
			data: file.code,
			mode: 420
		});
	}
	await nft({
		...args,
		localBuildFiles,
		files,
		span: rolldownSpan ?? new Span({ name: "vc.builder.backends.nft" }),
		ignoreNodeModules: true,
		ignore: args.config.excludeFiles
	});
	if (!handler) throw new Error(`Unable to resolve build handler for entrypoint: ${args.entrypoint}`);
	console.log(Colors.gray(`${Colors.bold(Colors.cyan("✓"))} Build complete — Using ${Colors.bold(args.entrypoint)} as the root entrypoint.`));
	return {
		files,
		handler,
		framework,
		localBuildFiles
	};
};

//#endregion
export { rolldown };
